
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/long-read-assembly/pages/day3c_phased_assemblies_in_action/">
      
      
        <link rel="prev" href="../day3b_assembly_cleanup_annotation/">
      
      
        <link rel="next" href="../../supplementary/supplementary_1/">
      
      <link rel="icon" href="../../theme_figures/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.18">
    
    
      
        <title>7. Phased Assemblies in Action - Long-read Assembly</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.26e3688c.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ecc896b0.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
        html.glightbox-open { overflow: initial; height: 100%; }
        .gslide-title { margin-top: 0px; user-select: text; }
        .gslide-desc { color: #666; user-select: text; }
        .gslide-image img { background: white; }
        
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
            </style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#7-phased-assemblies-in-action" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Long-read Assembly" class="md-header__button md-logo" aria-label="Long-read Assembly" data-md-component="logo">
      
  <img src="../../theme_figures/threelogo_fav.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Long-read Assembly
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              7. Phased Assemblies in Action
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
            
              <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_3" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
              </label>
            
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
            
              <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
              </label>
            
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_3">
            
              <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_2" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
              </label>
            
          
        </form>
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/long-read-assembly" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/long-read-assembly
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Long-read Assembly" class="md-nav__button md-logo" aria-label="Long-read Assembly" data-md-component="logo">
      
  <img src="../../theme_figures/threelogo_fav.png" alt="logo">

    </a>
    Long-read Assembly
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/long-read-assembly" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/long-read-assembly
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          Pages
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Pages
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day0_intro/" class="md-nav__link">
        1. Introduction - Long Read Assembly Workshop
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day1a_read_inputs/" class="md-nav__link">
        2. Sequence Data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day1b_familiarise_w_data/" class="md-nav__link">
        3. Familiarize Ourselves With The Data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day2a_assembly/" class="md-nav__link">
        4. Assembly
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day3a_assembly_qc/" class="md-nav__link">
        5. Assembly QC
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day3b_assembly_cleanup_annotation/" class="md-nav__link">
        6. Assembly Cleanup & Genome Annotation
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          7. Phased Assemblies in Action
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        7. Phased Assemblies in Action
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#visualizing-repeats-using-moddotplot" class="md-nav__link">
    Visualizing Repeats using ModDotPlot
  </a>
  
    <nav class="md-nav" aria-label="Visualizing Repeats using ModDotPlot">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#initial-setup" class="md-nav__link">
    Initial setup
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-chry-contigs-with-mashmap" class="md-nav__link">
    Find chrY contigs with MashMap
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#create-self-dot-plots-for-each-contig" class="md-nav__link">
    Create self dot plots for each contig
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
      
      
      
        
      
      <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Supplementary
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Supplementary
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_1/" class="md-nav__link">
        NeSI HPC Auth.Factors Setup & Jupyter Login
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_2/" class="md-nav__link">
        NeSI File system, Working directory and Symlinks
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_3/" class="md-nav__link">
        Jupyter Virtual Desktop
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_4/" class="md-nav__link">
        Additional materials
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#visualizing-repeats-using-moddotplot" class="md-nav__link">
    Visualizing Repeats using ModDotPlot
  </a>
  
    <nav class="md-nav" aria-label="Visualizing Repeats using ModDotPlot">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#initial-setup" class="md-nav__link">
    Initial setup
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-chry-contigs-with-mashmap" class="md-nav__link">
    Find chrY contigs with MashMap
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#create-self-dot-plots-for-each-contig" class="md-nav__link">
    Create self dot plots for each contig
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="7-phased-assemblies-in-action">7. Phased Assemblies in Action<a class="headerlink" href="#7-phased-assemblies-in-action" title="Permanent link">&para;</a></h1>
<p>What can you do with a phased assembly that you cannot do with a squashed
assembly? Why does having a highly-continuous and highly-accurate assembly
matter? In same cases, you don&rsquo;t need an assembly like that, e.g., if
doing targeted sequencing of a region of interest that fits inside the length
of a read. Conversely, some applications <em>require</em> a good assembly, e.g., when
studying repetitive DNA, especially long and/or interspersed repeats. Phasing
is also helpful when the haplotypes differ, especially when those differences
are structural in nature (e.g., copy number variants or large insertions,
deletions, duplications, or inversions). Let&rsquo;s take a look at the repeats
in chromosome Y of our HG002 assembly.</p>
<h2 id="visualizing-repeats-using-moddotplot">Visualizing Repeats using ModDotPlot<a class="headerlink" href="#visualizing-repeats-using-moddotplot" title="Permanent link">&para;</a></h2>
<p><a href="https://github.com/marbl/ModDotPlot">ModDotPlot</a> <!--(<a href="https://doi.org/##############">Sweeten klzzwxh:0004 2023</a>)-->
is a tool for visualizing tandem repeats using dot plots, similar to
<a href="https://mrvollger.github.io/StainedGlass">StainedGlass</a>
(<a href="https://doi.org/10.1093/bioinformatics/btac018">Vollger <em>et al.</em> 2022</a>), but
using sketching methods to radically reduce the computational requirements.
Visualizing tandem repeats like this <em>requires</em> an assembly that spans all the
repeat arrays and assembles them accurately, making this a potent example of
the benefits of combining highly-accurate, long reads (like PacBio HiFi) with
ultralong reads from ONT in the assembly process with 
<a href="https://github.com/chhylp123/hifiasm">hifiasm</a>
(<a href="https://doi.org/10.1038/s41587-022-01261-x">Cheng <em>et al.</em> 2022</a>) or
<a href="https://github.com/marbl/verkko">Verkko</a>
(<a href="https://doi.org/10.1038/s41587-023-01662-6">Rautiainen <em>et al.</em> 2023</a>). Take
a moment, and consider that. You <em>can&rsquo;t</em> do this with reads alone (even
long reads). Mapping to a good reference (e.g., CHM13-T2T) for your species of
interest (if one exists) <em>won&rsquo;t</em> work either because the alignment
software can&rsquo;t distinguish between repeat copies.</p>
<details class="clipboard-question">
<summary>What are sketching methods ?</summary>
<p>Sketching is a technique to create reduced-representations
of a sequence. The most widely-known option for sketching is probably
minimizers, made particularly popular with tools like
<a href="https://lh3.github.io/minimap2">minimap2</a>
(<a href="https://doi.org/10.1093/bioinformatics/bty191">Li 2018</a>),
which applies minimizers to the alignment problem. We discussed
minimizers earlier when running MashMap. The minimizer for any given
window is the lexicographically smallest canonical k-mer. Several
variants to minimizers exist, e.g., syncmers, minmers, and modimizers,
the latter of which is used in ModDotPlot. Each sketching method has
different properties depending on how they select the subsequence used
to represent a larger area, whether they allow overlaps, whether a
certain density of representative sequences is enforced in any given
window, whether the neighboring windows are dependent on eachother, etc.
In general, the representative sequences are found by sliding along the
sequence and selecting a representative subsequence in the given window.</p>
<p>Many other tools use sketching in some way, here are a few examples:</p>
<ul>
<li><a href="https://github.com/marbl/Mash">Mash</a> (<a href="https://doi.org/10.1186/s13059-016-0997-x">Ondov <em>et al.</em> 2016</a>)</li>
<li><a href="https://github.com/marbl/MashMap">MashMap</a> (<a href="https://doi.org/10.1101/2023.05.16.540882">Kille <em>et al.</em> 2023</a>)</li>
<li>
<p><a href="https://github.com/maickrau/MBG">MBG</a> (<a href="https://doi.org/10.1093/bioinformatics/btab004">Rautiainen &amp; Marschall 2021</a>)
            <strong>
                &lt;-- Used in
                <a href="https://github.com/marbl/verkko">Verkko</a></strong>
                (<a href="https://doi.org/10.1038/s41587-023-01662-6">Rautiainen <em>et al.</em> 2023</a>)!</p>
</li>
<li>
<p><a href="https://github.com/chhylp123/hifiasm">hifiasm</a> (<a href="https://doi.org/10.1038/s41587-022-01261-x">Cheng <em>et al.</em> 2022</a>)</p>
</li>
</ul>
</details>
<p>We're going to run ModDotPlot on part of the Y chromosome from our
earlier assembly. First we&rsquo;ll need to identify the appropriate chunk,
which can be done a few different ways depending on the sequence you&rsquo;re
looking for. ChrY somtimes appears tied on only one end to chrX in the Bandage
plot, making it easy to identify the Y node (which is shorter than the longer X
node). This method may not get all of the Y chromosome because it may be in
separate pieces (i.e., it isn&rsquo;t T2T), and we often see chrX and chrY
untangled from eachother in one or two pieces each, so topology of the graph
isn&rsquo;t always a reliable option (though, we recommend labelling the graph
with the chromosome names to get a feel for how your assembly looks). The ideal
way is to map some known sequence against your assembly or to map your assembly
against a known reference. Since we&rsquo;re using data from HG002 (a human), we
can map against the human CHM13-T2T reference.</p>
<h3 id="initial-setup">Initial setup<a class="headerlink" href="#initial-setup" title="Permanent link">&para;</a></h3>
<p><strong>Make a directory to work in</strong></p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="nb">cd</span><span class="w"> </span>~/lra
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>mkdir<span class="w"> </span>day3c-moddotplot
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="nb">cd</span><span class="w"> </span>day3c-moddotplot
</code></pre></div>
</div>
<p><strong>Get the files</strong></p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/chm13/chm13v2.0.fa<span class="w"> </span>chm13.fa
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/assembly.haplotype2.fasta<span class="w"> </span>hg002.hap2.fa
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/assembly.haplotype2.fasta.fai<span class="w"> </span>hg002.hap2.fa.fai
</code></pre></div>
</div>
<p>Note that we&rsquo;re cheating a bit. We could map against the entire diploid
assembly, but we&rsquo;ve already determined that chrY is in haplotype2, so
we&rsquo;ll save some time and work with only that haplotype.</p>
<h3 id="find-chry-contigs-with-mashmap">Find chrY contigs with MashMap<a class="headerlink" href="#find-chry-contigs-with-mashmap" title="Permanent link">&para;</a></h3>
<p><strong>Do the alignments</strong></p>
<p>We&rsquo;ll do the alignments with <a href="https://github.com/marbl/MashMap">MashMap</a>
(<a href="https://doi.org/10.1101/2023.05.16.540882">Kille <em>et al.</em> 2023</a>). This should
take ~25 minutes with 4 CPUs and should use &lt;4 GB RAM. With 16 CPUs, it
should take ~4 minutes and will use &lt;8 GB RAM. This command is the same as
the one you ran earlier, except that this time it is on haplotype2 and the
percent identity threshold is lower, which should help us recruit alignments in
the satellites on chrY. We&rsquo;ll use pre-computed results to avoid needing
to wait for the job to complete, but this is what the command would look like:</p>
<div class="admonition terminal">
<p class="admonition-title">view, but do not run this code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>module<span class="w"> </span>purge
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>module<span class="w"> </span>load<span class="w"> </span>MashMap/3.0.4-Miniconda3
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>mashmap<span class="w"> </span>-f<span class="w"> </span><span class="s2">&quot;one-to-one&quot;</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="w">    </span>-k<span class="w"> </span><span class="m">16</span><span class="w"> </span>--pi<span class="w"> </span><span class="m">90</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="w">    </span>-s<span class="w"> </span><span class="m">100000</span><span class="w"> </span>-t<span class="w"> </span><span class="m">4</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a><span class="w">    </span>-r<span class="w"> </span>chm13.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a><span class="w">    </span>-q<span class="w"> </span>hg002.hap2.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a><span class="w">    </span>-o<span class="w"> </span>hg2hap2-x-chm13.ssv
</code></pre></div>
</div>
<details class="question">
<summary>How would I submit this as a job?</summary>
<p>Submit it as a job with <code>sbatch</code>. First copy the command into a
script named <code>mashmap.sh</code>:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="c1">#SBATCH --job-name      mashmap</span>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="c1">#SBATCH --cpus-per-task 16</span>
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a><span class="c1">#SBATCH --time          00:20:00</span>
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a><span class="c1">#SBATCH --mem           8G</span>
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a><span class="c1">#SBATCH --partition     milan</span>
<a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a><span class="c1">#SBATCH --output        slurmlogs/%x.%j.out</span>
<a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a><span class="c1">#SBATCH --error         slurmlogs/%x.%j.err</span>
<a id="__codelineno-3-11" name="__codelineno-3-11" href="#__codelineno-3-11"></a>
<a id="__codelineno-3-12" name="__codelineno-3-12" href="#__codelineno-3-12"></a>
<a id="__codelineno-3-13" name="__codelineno-3-13" href="#__codelineno-3-13"></a>module<span class="w"> </span>purge
<a id="__codelineno-3-14" name="__codelineno-3-14" href="#__codelineno-3-14"></a>module<span class="w"> </span>load<span class="w"> </span>MashMap/3.0.4-Miniconda3
<a id="__codelineno-3-15" name="__codelineno-3-15" href="#__codelineno-3-15"></a>
<a id="__codelineno-3-16" name="__codelineno-3-16" href="#__codelineno-3-16"></a>mashmap<span class="w"> </span>-f<span class="w"> </span><span class="s2">&quot;one-to-one&quot;</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-17" name="__codelineno-3-17" href="#__codelineno-3-17"></a><span class="w">    </span>-k<span class="w"> </span><span class="m">16</span><span class="w"> </span>--pi<span class="w"> </span><span class="m">90</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-18" name="__codelineno-3-18" href="#__codelineno-3-18"></a><span class="w">    </span>-s<span class="w"> </span><span class="m">100000</span><span class="w"> </span>-t<span class="w"> </span><span class="m">16</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-19" name="__codelineno-3-19" href="#__codelineno-3-19"></a><span class="w">    </span>-r<span class="w"> </span>chm13.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-20" name="__codelineno-3-20" href="#__codelineno-3-20"></a><span class="w">    </span>-q<span class="w"> </span>hg002.hap2.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-3-21" name="__codelineno-3-21" href="#__codelineno-3-21"></a><span class="w">    </span>-o<span class="w"> </span>hg2hap2-x-chm13.ssv
</code></pre></div>
<p>Then submit the job with the following command:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>sbatch<span class="w"> </span>mashmap.sh
</code></pre></div>
</div>
</details>
<p><strong>Get the pre-computed results</strong></p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>cp<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/hap2-mashmap/hg2hap2-x-chm13.ssv<span class="w"> </span>./
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>cp<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/hap2-mashmap/mashmap.log<span class="w"> </span>./
</code></pre></div>
</div>
<p><strong>View the output file</strong></p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>less<span class="w"> </span>-S<span class="w"> </span>hg2hap2-x-chm13.ssv
</code></pre></div>
</div>
<p>There is more information present than we need, and we can simplify things by
looking at long alignments only.</p>
<details class="info">
<summary>Optionally, view the MashMap log file</summary>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>less<span class="w"> </span>-S<span class="w"> </span>mashmap.log
</code></pre></div>
</div>
</details>
<p><strong>Subset the alignments</strong></p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>awk<span class="w"> </span><span class="s1">&#39;BEGIN{FS=&quot; &quot;; OFS=FS}{if($6 == &quot;chrY&quot; &amp;&amp; ($4-$3) &gt;= 1000000){print $0}}&#39;</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="w">    </span>&lt;<span class="w"> </span>hg2hap2-x-chm13.ssv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a><span class="w">    </span>&gt;<span class="w"> </span>hg2hap2-x-chm13.gt1m-chrY.ssv
</code></pre></div>
</div>
<details class="clipboard-question">
<summary>What is the <code>awk</code> command doing ?</summary>
<p>This <code>awk</code> command is keeping only alignments (remember, one
alignment is on each line of the file) that map to chrY; the sixth column
is the "reference" or "target" sequence name. The third and fourth columns
are respectively the start and stop positions of the aligned region on the
"query" sequence (i.e., a contig from our assembly); thus, we&rsquo;re
keeping only alignments that are 1 Mbp or longer. This also has the
consequence of ignoring contigs that are shorter than 1 Mbp.</p>
<details class="info">
<summary>Wait, what are each of the columns again?</summary>
<div class="admonition quote">
<p class="admonition-title">According to the MashMap README</p>
<p>The output is space-delimited with each line consisting of query name, length, 0-based start, end, strand, target name, length, start, end, and mapping nucleotide identity.</p>
</div>
<p>Re-written as a numbered list:</p>
<div class="admonition rectangle-list">
<p class="admonition-title">MashMap output columns</p>
<ol>
<li>query name</li>
<li>length</li>
<li>0-based start</li>
<li>end</li>
<li>strand</li>
<li>target name</li>
<li>length</li>
<li>start</li>
<li>end</li>
<li>mapping nucleotide identity</li>
</ol>
</div>
</details>
</details>
<p><strong>View the output file</strong></p>
<p>Now that we&rsquo;ve culled the alignments, viewing them should be much easier:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>less<span class="w"> </span>-S<span class="w"> </span>hg2hap2-x-chm13.gt1m-chrY.ssv
</code></pre></div>
</div>
<p><strong>Which contigs belong to chrY?</strong></p>
<details class="success">
<summary>Answer</summary>
<p><code>pat-0000724</code>, <code>pat-0000725</code>,  and <code>pat-0000727</code> are probably chrY.</p>
</details>
<p>Others may be as well, but it is difficult to tell without a more refined
investigation, and this is sufficient for our purposes.</p>
<p><strong>How can we tell?</strong></p>
<p>Anything with a long, high-identity alignment is a pretty good candidate,
especially if the contig was determined to be paternal using trio markers (as
these were). One thing that may help is seeing what percentage of the contig is
covered by each alignment:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>awk<span class="w"> </span><span class="s1">&#39;BEGIN{FS=&quot; &quot;; OFS=&quot;\t&quot;; print &quot;Contig&quot;, &quot;Length&quot;, &quot;Percent Identity&quot;, &quot;Percent Aligned&quot;}{print $1, $2, $10 &quot;%&quot;, ($4-$3)/$2*100 &quot;%&quot;}&#39;</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="w">    </span>hg2hap2-x-chm13.gt1m-chrY.ssv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="w">    </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-ts<span class="w"> </span><span class="s1">$&#39;\t&#39;</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="w">    </span>&gt;<span class="w"> </span>hg2hap2-x-chm13.gt1m-chrY.annotated.txt
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a>less<span class="w"> </span>-S<span class="w"> </span>hg2hap2-x-chm13.gt1m-chrY.annotated.txt
</code></pre></div>
</div>
<!-- this was wrong- it was based on only --pi 95, which is why we saw only a small amount:
The astute observer will wonder why pat-0000724 should be included when
it has only 1% of the contig is aligned. Note that only 1% of the
contig is aligned in a block of &gt;= 1 Mbp. Let&rsquo;s add back in
the short alignments for these three contigs and then look at the
percentages for each contig:

<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>awk &#39;BEGIN{FS=&quot; &quot;; OFS=FS}{if($6 == &quot;chrY&quot; &amp;&amp; $1 ~ /^pat-000042[457]$/){print $0}}&#39; \
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>    hg2hap2-x-chm13.ssv \
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a>    &gt; hg2hap2-x-chm13.selected.ssv
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a>
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a>awk &#39;BEGIN{FS=&quot; &quot;; OFS=&quot;\t&quot;; print &quot;Contig&quot;, &quot;Length&quot;, &quot;Percent Identity&quot;, &quot;Percent Aligned&quot;}{print $1, $2, $10 &quot;%&quot;, ($4-$3)/$2*100 &quot;%&quot;}&#39; \
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a>    hg2hap2-x-chm13.selected.ssv \
<a id="__codelineno-11-7" name="__codelineno-11-7" href="#__codelineno-11-7"></a>    | column -ts $&#39;\t&#39; \
<a id="__codelineno-11-8" name="__codelineno-11-8" href="#__codelineno-11-8"></a>    &gt; hg2hap2-x-chm13.selected.annotated.txt
<a id="__codelineno-11-9" name="__codelineno-11-9" href="#__codelineno-11-9"></a>
<a id="__codelineno-11-10" name="__codelineno-11-10" href="#__codelineno-11-10"></a>less -S hg2hap2-x-chm13.selected.annotated.txt
</code></pre></div>

To prevent you from having to do the math in your head, here are the
percentages of each contig aligned to chrY (sum of the final column in
the `hg2hap2-x-chm13.selected.annotated.txt` file):
-->

<p>To prevent you from having to do the math in your head, here are the
percentages of each contig aligned to chrY (sum of the final column in
the <code>hg2hap2-x-chm13.gt1m-chrY.annotated.txt</code> file):</p>
<table>
    <thead>
        <td>
            HG002 Contig
        </td>
        <td>
            % aligned to chrY
        </td>
    </thead>
    <tr>
        <td>
            pat-0000724
        </td>
        <td>
            84.332%
        </td>
    </tr>
    <tr>
        <td>
            pat-0000725
        </td>
        <td>
            99.9999%
        </td>
    </tr>
    <tr>
        <td>
            pat-0000727
        </td>
        <td>
            98.342%
        </td>
    </tr>
</table>

<p>More short contigs may also belong to chrY, but we would need to investigate
more carefully to find them. We would want to confirm whether telomeres are
present on the end of the terminal contigs. Also, we do not expect perfect or
complete alignments here; not only are we using a sketching-based alignment
method, but any aligner would likely struggle when aligning repetitve DNA. So,
how much of chrY did we capture with these three contigs? You can see that
these three contigs cover the majority, if not all, of chrY:</p>
<p><a class="glightbox" href="https://github.com/human-pangenomics/hprc-tutorials/blob/GA-workshop/assembly/genomics_aotearoa/images/assembly-in-action/mashmap_hg002-x-chm13_chrY.png?raw=true" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img src="https://github.com/human-pangenomics/hprc-tutorials/blob/GA-workshop/assembly/genomics_aotearoa/images/assembly-in-action/mashmap_hg002-x-chm13_chrY.png?raw=true" alt="Dotplot of HG002 contigs against CHM13 chrY"></a></p>
<h2 id="create-self-dot-plots-for-each-contig">Create self dot plots for each contig<a class="headerlink" href="#create-self-dot-plots-for-each-contig" title="Permanent link">&para;</a></h2>
<p>ModDotPlot is still in development, and it cannot currently support doing
multiple self comparisons at one time. We will need to create separate fasta
files for our contigs of interest.</p>
<p><strong>Create and index subset fastas</strong></p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>module<span class="w"> </span>load<span class="w"> </span>SAMtools/1.16.1-GCC-11.3.0
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a><span class="k">for</span><span class="w"> </span>CTG<span class="w"> </span><span class="k">in</span><span class="w"> </span>pat-000072<span class="o">{</span><span class="m">4</span>,5,7<span class="o">}</span>
<a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a><span class="k">do</span>
<a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a><span class="w">    </span>samtools<span class="w"> </span>faidx<span class="w"> </span>hg002.hap2.fa<span class="w"> </span><span class="si">${</span><span class="nv">CTG</span><span class="si">}</span><span class="w"> </span>&gt;<span class="w"> </span>hg002.hap2.<span class="si">${</span><span class="nv">CTG</span><span class="si">}</span>.fa
<a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a><span class="w">    </span>samtools<span class="w"> </span>faidx<span class="w"> </span>hg002.hap2.<span class="si">${</span><span class="nv">CTG</span><span class="si">}</span>.fa
<a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a><span class="k">done</span>
</code></pre></div>
</div>
<p><strong>Run ModDotPlot</strong></p>
<p>On sequences of this size, ModDotPlot is relatively quick. It has a reasonable
memory footprint for sequences &lt;10 Mbp, but memory usage can exceed 20 GB for
large sequences (&gt;100 Mbp). Let's create a script to submit with
<code>sbatch</code>. Paste the following into <code>moddotplot.sh</code></p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="ch">#!/bin/bash -e </span>
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="c1">#SBATCH --job-name      moddotplot</span>
<a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="c1">#SBATCH --cpus-per-task 4</span>
<a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a><span class="c1">#SBATCH --time          00:10:00</span>
<a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="c1">#SBATCH --mem           8G</span>
<a id="__codelineno-13-8" name="__codelineno-13-8" href="#__codelineno-13-8"></a><span class="c1">#SBATCH --partition     milan</span>
<a id="__codelineno-13-9" name="__codelineno-13-9" href="#__codelineno-13-9"></a><span class="c1">#SBATCH --output        slurmlogs/%x.%j.log</span>
<a id="__codelineno-13-10" name="__codelineno-13-10" href="#__codelineno-13-10"></a>
<a id="__codelineno-13-11" name="__codelineno-13-11" href="#__codelineno-13-11"></a>module<span class="w"> </span>purge
<a id="__codelineno-13-12" name="__codelineno-13-12" href="#__codelineno-13-12"></a>module<span class="w"> </span>load<span class="w"> </span>ModDotPlot/2023-06-gimkl-2022a-Python-3.11.3
<a id="__codelineno-13-13" name="__codelineno-13-13" href="#__codelineno-13-13"></a>
<a id="__codelineno-13-14" name="__codelineno-13-14" href="#__codelineno-13-14"></a><span class="k">for</span><span class="w"> </span>CTG<span class="w"> </span><span class="k">in</span><span class="w"> </span>pat-000072<span class="o">{</span><span class="m">4</span>,5,7<span class="o">}</span>
<a id="__codelineno-13-15" name="__codelineno-13-15" href="#__codelineno-13-15"></a><span class="k">do</span>
<a id="__codelineno-13-16" name="__codelineno-13-16" href="#__codelineno-13-16"></a><span class="w">    </span>moddotplot<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-17" name="__codelineno-13-17" href="#__codelineno-13-17"></a><span class="w">        </span>-k<span class="w"> </span><span class="m">21</span><span class="w"> </span>-id<span class="w"> </span><span class="m">85</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-18" name="__codelineno-13-18" href="#__codelineno-13-18"></a><span class="w">        </span>-i<span class="w"> </span>hg002.hap2.<span class="si">${</span><span class="nv">CTG</span><span class="si">}</span>.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-19" name="__codelineno-13-19" href="#__codelineno-13-19"></a><span class="w">        </span>-o<span class="w"> </span>mdp_hg002-<span class="si">${</span><span class="nv">CTG</span><span class="si">}</span>
<a id="__codelineno-13-20" name="__codelineno-13-20" href="#__codelineno-13-20"></a><span class="k">done</span>
</code></pre></div>
</div>
<p>Then submit with <code>sbatch</code>:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>sbatch<span class="w"> </span>moddotplot.sh
</code></pre></div>
</div>
<details class="clipboard-question">
<summary>What do these parameters do ?</summary>
<p>You can run <code>moddotplot -h</code> to find out (and enjoy some excellent ASCII art).
Here are the options we used:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>Required input:
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>  -i INPUT [INPUT ...], --input INPUT [INPUT ...]
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>                        Path to input fasta file(s)
<a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a>
<a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a>Mod.Plot distance matrix commands:
<a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a>  -k KMER, --kmer KMER  k-mer length. Must be &lt; 32 (default: 21)
<a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a>
<a id="__codelineno-15-8" name="__codelineno-15-8" href="#__codelineno-15-8"></a>  -id IDENTITY, --identity IDENTITY
<a id="__codelineno-15-9" name="__codelineno-15-9" href="#__codelineno-15-9"></a>                        Identity cutoff threshold. (default: 80)
<a id="__codelineno-15-10" name="__codelineno-15-10" href="#__codelineno-15-10"></a>
<a id="__codelineno-15-11" name="__codelineno-15-11" href="#__codelineno-15-11"></a>  -o OUTPUT, --output OUTPUT
<a id="__codelineno-15-12" name="__codelineno-15-12" href="#__codelineno-15-12"></a>                        Name for bed file and plots. Will be set to input fasta file name if not provided. (default: None)
</code></pre></div>
</details>
<p><strong>Inspect the output files</strong></p>
<p>First, take a look at the log file:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>less<span class="w"> </span>-S<span class="w"> </span>slurmlogs/moddotplot.*.log
</code></pre></div>
</div>
<p>Then note that for every run, we created <code>_HIST.png</code> and <code>_TRI.png</code> files. The
HIST files show the distribution of the Percent Identity of the alignments. The
TRI files show everything above (or below, depending on how you look at it) the
diagonal of the dotplot, rotated such that the diagonal is along the X-axis of
the plot. Go ahead a view these files now.</p>
<div class="admonition question">
<p class="admonition-title">What do you observe?</p>
</div>





                
              </article>
            </div>
          
          
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
            Back to top
          </button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../day3b_assembly_cleanup_annotation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 6. Assembly Cleanup &amp; Genome Annotation" rel="prev">
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                6. Assembly Cleanup & Genome Annotation
              </div>
            </div>
          </a>
        
        
          
          <a href="../../supplementary/supplementary_1/" class="md-footer__link md-footer__link--next" aria-label="Next: NeSI HPC Auth.Factors Setup &amp; Jupyter Login" rel="next">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                NeSI HPC Auth.Factors Setup & Jupyter Login
              </div>
            </div>
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Long-read Assembly workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.74e28a9f.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.220ee61c.min.js"></script>
      
        
          <script src="../../javascripts/mathjax.js"></script>
        
      
        
          <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
        
      
        
          <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        
      
    
  <script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>