<center>![image](./theme_figures/institution_logos.png){width="500"}</center>
<hr style="border:1px solid grey">
<center>
# **Long-read Assembly**
</center>

<center>
![image](./theme_figures/double_helix_mirror.png){width="300"}
</center>


!!! calendar-days "Workshop schedule"

    | Day                 | Time    | Lesson overview                                                                 |
    | :------------------ | :------ | :----------------------------------------------------------------------- |
    | Tuesday, 4th July   | 1:00PM <br> 2:00PM <br> 3:00PM | [Introduction to the workshop](pages/day0_intro.md) <br> [How are assemblies created?](pages/day1a_read_inputs.md) <br> [Familiarize Ourselves with the Data](pages/day1b_familiarise_w_data.md) <br> Talk by Assistant Professor Karen Miga |
    | Wednesday, 5th July | 9:00AM <br> 12:00PM <br> 1:00PM <br> 2:00PM <br> 3:00PM | [Assembly](pages/day2a_assembly.md) <br> Lunch break <br> [Assembly continued](pages/day2a_assembly.md) <br> Talk by Dr Brandon Pickett <br> Talk by Dr Joseph Guhlin                                      |
    | Thursday, 6th July  | 9:00AM <br> 10:00AM <br> 11:00AM <br> 12:00PM <br> 1:00PM  | [Assembly QC](pages/day3a_assembly_qc.md) <br> [Assembly clean-up](pages/day3b_assembly_cleanup_annotation.md) <br> [Phased assemblies in action](pages/day3c_phased_assemblies_in_action.md) <br> Lunch break <br> Talk by Assistant Professor Benedict Paten                              |
  



??? example "Supplementary materials"

    | Supplementary   | Topic                                                                                      |
    | :-------------- | :----------------------------------------------------------------------------------------- |
    | Supplementary 1 | [NeSI HPC Authentication Factor Setup and Jupyter Login](supplementary/supplementary_1.md) |
    | Supplementary 2 | [NeSI File System, Working Directory, and Symlinks](supplementary/supplementary_2.md) |
    | Supplementary 3 | [Jupyter Virtual Desktop](supplementary/supplementary_3.md)|


!!! copyright "Attribution notice"

    - Material used to prepare for the workshop was extracted from https://github.com/human-pangenomics/hprc-tutorials/tree/GA-workshop
 

!!! book-atlas "References"

    - Lorig-Roach, Ryan, et al. "Phased nanopore assembly with Shasta and modular graph phasing with GFAse." *bioRxiv* (2023): 2023-02.

    - Porubsky, David, et al. "Fully phased human genome assembly without parental data using single-cell strand sequencing and long reads." *Nature biotechnology* 39.3 (2021): 302-308.

    - Garg, Shilpa. "Towards routine chromosome-scale haplotype-resolved reconstruction in cancer genomics." *Nature Communications* 14.1 (2023): 1358.

    - Lorig-Roach, Ryan, et al. "Phased nanopore assembly with Shasta and modular graph phasing with GFAse." *bioRxiv* (2023): 2023-02.

    - Porubsky, David, et al. "Fully phased human genome assembly without parental data using single-cell strand sequencing and long reads." *Nature biotechnology* 39.3 (2021): 302-308.

    - Rautiainen, Mikko, et al. "Telomere-to-telomere assembly of diploid chromosomes with Verkko." *Nature Biotechnology* (2023): 1-9.

    - Rhie, Arang, et al. "Merqury: reference-free quality, completeness, and phasing assessment for genome assemblies." *Genome biology* 21.1 (2020): 1-27.