
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/long-read-assembly/pages/day3a_assembly_qc/">
      
      
        <link rel="prev" href="../day2a_assembly/">
      
      
        <link rel="next" href="../day3b_assembly_cleanup_annotation/">
      
      <link rel="icon" href="../../theme_figures/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.18">
    
    
      
        <title>5. Assembly QC - Long-read Assembly</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.26e3688c.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ecc896b0.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
        html.glightbox-open { overflow: initial; height: 100%; }
        .gslide-title { margin-top: 0px; user-select: text; }
        .gslide-desc { color: #666; user-select: text; }
        .gslide-image img { background: white; }
        
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
            </style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#5-assembly-qc" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Long-read Assembly" class="md-header__button md-logo" aria-label="Long-read Assembly" data-md-component="logo">
      
  <img src="../../theme_figures/threelogo_fav.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Long-read Assembly
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              5. Assembly QC
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
            
              <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_3" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
              </label>
            
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
            
              <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
              </label>
            
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_3">
            
              <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_2" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
              </label>
            
          
        </form>
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/long-read-assembly" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/long-read-assembly
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Long-read Assembly" class="md-nav__button md-logo" aria-label="Long-read Assembly" data-md-component="logo">
      
  <img src="../../theme_figures/threelogo_fav.png" alt="logo">

    </a>
    Long-read Assembly
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/long-read-assembly" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/long-read-assembly
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          Pages
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Pages
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day0_intro/" class="md-nav__link">
        1. Introduction - Long Read Assembly Workshop
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day1a_read_inputs/" class="md-nav__link">
        2. Sequence Data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day1b_familiarise_w_data/" class="md-nav__link">
        3. Familiarize Ourselves With The Data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day2a_assembly/" class="md-nav__link">
        4. Assembly
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          5. Assembly QC
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        5. Assembly QC
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#contiguity-assembly-statistics-using-gfastats" class="md-nav__link">
    Contiguity (assembly statistics using gfastats)
  </a>
  
    <nav class="md-nav" aria-label="Contiguity (assembly statistics using gfastats)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-gfastats-on-a-fasta" class="md-nav__link">
    Run gfastats on a FASTA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-gfastats-on-a-gfa" class="md-nav__link">
    Run gfastats on a GFA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compare-two-graphs-stats" class="md-nav__link">
    Compare two graphs' stats
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#correctness-qv-using-merqury" class="md-nav__link">
    Correctness (QV using Merqury)
  </a>
  
    <nav class="md-nav" aria-label="Correctness (QV using Merqury)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#running-meryl-and-genomescope-on-the-e-coli-verkko-assembly" class="md-nav__link">
    Running Meryl and GenomeScope on the E. coli Verkko assembly
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-merqury-in-solo-mode" class="md-nav__link">
    Using Merqury in solo mode
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-merqury-in-trio-mode" class="md-nav__link">
    Using Merqury in trio mode
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#what-do-different-phasing-approaches-look-like-in-merqury" class="md-nav__link">
    What do different phasing approaches look like in Merqury?
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#switch-and-hamming-errors-using-yak" class="md-nav__link">
    Switch and Hamming errors using yak
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#completeness-asmgene" class="md-nav__link">
    Completeness (asmgene)
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day3b_assembly_cleanup_annotation/" class="md-nav__link">
        6. Assembly Cleanup & Genome Annotation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../day3c_phased_assemblies_in_action/" class="md-nav__link">
        7. Phased Assemblies in Action
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
      
      
      
        
      
      <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Supplementary
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Supplementary
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_1/" class="md-nav__link">
        NeSI HPC Auth.Factors Setup & Jupyter Login
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_2/" class="md-nav__link">
        NeSI File system, Working directory and Symlinks
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_3/" class="md-nav__link">
        Jupyter Virtual Desktop
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/supplementary_4/" class="md-nav__link">
        Additional materials
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#contiguity-assembly-statistics-using-gfastats" class="md-nav__link">
    Contiguity (assembly statistics using gfastats)
  </a>
  
    <nav class="md-nav" aria-label="Contiguity (assembly statistics using gfastats)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-gfastats-on-a-fasta" class="md-nav__link">
    Run gfastats on a FASTA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-gfastats-on-a-gfa" class="md-nav__link">
    Run gfastats on a GFA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compare-two-graphs-stats" class="md-nav__link">
    Compare two graphs' stats
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#correctness-qv-using-merqury" class="md-nav__link">
    Correctness (QV using Merqury)
  </a>
  
    <nav class="md-nav" aria-label="Correctness (QV using Merqury)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#running-meryl-and-genomescope-on-the-e-coli-verkko-assembly" class="md-nav__link">
    Running Meryl and GenomeScope on the E. coli Verkko assembly
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-merqury-in-solo-mode" class="md-nav__link">
    Using Merqury in solo mode
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-merqury-in-trio-mode" class="md-nav__link">
    Using Merqury in trio mode
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#what-do-different-phasing-approaches-look-like-in-merqury" class="md-nav__link">
    What do different phasing approaches look like in Merqury?
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#switch-and-hamming-errors-using-yak" class="md-nav__link">
    Switch and Hamming errors using yak
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#completeness-asmgene" class="md-nav__link">
    Completeness (asmgene)
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="5-assembly-qc">5. Assembly QC<a class="headerlink" href="#5-assembly-qc" title="Permanent link">&para;</a></h1>
<p>Now that we have understood our data types (day 1) and put them through an assembly algorithm (day 2), we have this file of A's, T's, C's, and G's that's supposed to be our assembly. This file is meant to represent a biological reality, so let's try to assess its quality through several lens, some biological and some more technical. One way to remember the ways we evaluate assemblies is by thinking about the "3C's": <strong>contiguity</strong>, <strong>correctness</strong>, and <strong>completeness</strong>.</p>
<div class="admonition question">
<p class="admonition-title">Food for thought</p>
<ul>
<li>What do you think a 'good' <em>de novo</em> assembly looks like?</li>
<li>What are some qualities of an assembly that you might be interested in measuring?</li>
</ul>
</div>
<h2 id="contiguity-assembly-statistics-using-gfastats">Contiguity (assembly statistics using gfastats)<a class="headerlink" href="#contiguity-assembly-statistics-using-gfastats" title="Permanent link">&para;</a></h2>
<p>Recall that the sequences in our assembly are referred to as <em>contigs</em>.</p>
<p>Normally, when we receive a hodgepodge of things with different values of the same variable, such as our contigs of varying lengths, we are inclined to use descriptive statistics such as average or median to try to get a grasp on how our data looks. However, it can be hard to compare average contig length between assemblies&mdash;if they have the same total size and same number of contigs, it's still the same average, even if it's five contigs of 100bp, or one 460 bp contig and four 10bp ones! This matters for assembly because ideally we want <em>fewer</em> contigs that are <em>larger</em>. </p>
<p>Median comes closer to reaching what we're trying to measure, but it can be skewed by having many very small contigs, so instead a popular metric for assessing assemblies is <em>N50</em>.</p>
<p>The N50 is similar to the median in that one must first sort the numbers, but then instead of taking the middle value, the N50 value is the <em>length of the first contig that is equal to or greater than half of the assembly sum</em>. But that can be hard to understand verbally, so let's look at it visually:</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/N50.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="N50 schematic" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/N50.png" /></a>
<em>Image adapted from <a href='https://www.molecularecologist.com/2017/03/29/whats-n50/'>Elin Videvall at The Molecular Ecologist</a>.</em> </p>
<p>The N50 can be interpreted as such: given an N50 value, 50% of the sequence in that assembly is contained in contigs of that length or longer. Thus, N50 has been traditionally used as the assembly statistic of choice for comparing assemblies, as it's more intuitive (compared to average contig length) to see that an assembly with an N50 value of 100Mbp is more contiguous than one with an N50 value of 50MBp, since it seems like there are more larger contigs in the former assembly.</p>
<p>Another statistic that is often reported with N50 is the <em>L50</em>, which is the rank of the contig that gives the N50 value. For instance, in the above image, the L50 would be 3, because it would be the third largest contig that gives the N50 value. L50 is useful for contextualizing the N50, because it gives an idea of how many contigs make up that half of your assembly. </p>
<details class="note">
<summary>N50 or NG50?</summary>
<p>Another measurement you might see is NG50. This is the N50 value, just calculated using a given genome size instead of the sum of the contigs.</p>
</details>
<p>Given how the N50 value can be so affected by addition or removal of small contigs, another metric has come into use: <a href="https://lh3.github.io/2020/04/08/a-new-metric-on-assembly-contiguity">the area under the (N50) curve</a>, or the auN value. Though N50 is measured at the 50% mark, we could make similar values for any value of x, for instance N30 would be the value where 30% of the sequence in that assembly is of that length or longer. These metrics are thus called Nx statistics, and one could plot them against contig length to get an <em>Nx curve</em>, which gives a more nuanced view of the actual contig size distribution of your assembly. </p>
<p><center><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/NGx_plot.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="NGx plot" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/NGx_plot.png" /></a></center></p>
<p><center><small>Image adapted from <a href='https://lh3.github.io/2020/04/08/a-new-metric-on-assembly-contiguity'>Heng Li's blog, which in turn adapts it from a NIBS workshop</a></small></center></p>
<p>auN tries to capture the nature of this curve, instead of a value from an arbitrary point on it. On the above example, each step on the curve represents a contig (length on the y-axis), so the black curve is the most contiguous as it has one contig that covers over 40% of the assembly. Despite that, this assembly would have the same N50 value (on the x-axis) as multiple other assemblies that are more fragmented in the same area. </p>
<h3 id="run-gfastats-on-a-fasta">Run gfastats on a FASTA<a class="headerlink" href="#run-gfastats-on-a-fasta" title="Permanent link">&para;</a></h3>
<p>Let's get some basic statistics for an assembly using a tool called <strong>gfastats</strong>, which will output metrics such as N50, auN, total size, etc. We can try it out on a Verkko trio assembly of HG002 that's already been downloaded onto NeSI.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="c1">## let&#39;s symlink the file in a way that&#39;s easier to refer to</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="nb">cd</span><span class="w"> </span>~/lra
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>day3_assembly_qc/gfastats
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="nb">cd</span><span class="w"> </span>day3_assembly_qc/gfastats
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/assembly.*.fasta<span class="w"> </span>.
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>ls<span class="w"> </span>-la<span class="w">          </span><span class="c1"># you should see a bunch of &quot;files&quot; that are actually symlinks pointing to their actual content</span>
</code></pre></div>
Now we're ready to roll!
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>module purge
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>module load gfastats
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>gfastats assembly.haplotype1.fasta
</code></pre></div></p>
<ul>
<li>Shouldn't take about 10 seconds </li>
</ul>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>+++Assembly<span class="w"> </span>summary+++:<span class="w"> </span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="c1"># scaffolds: 81</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>Total<span class="w"> </span>scaffold<span class="w"> </span>length:<span class="w"> </span><span class="m">3025610697</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>Average<span class="w"> </span>scaffold<span class="w"> </span>length:<span class="w"> </span><span class="m">37353218</span>.48
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>Scaffold<span class="w"> </span>N50:<span class="w"> </span><span class="m">112270693</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>Scaffold<span class="w"> </span>auN:<span class="w"> </span><span class="m">120344050</span>.34
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>Scaffold<span class="w"> </span>L50:<span class="w"> </span><span class="m">10</span>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>Largest<span class="w"> </span>scaffold:<span class="w"> </span><span class="m">242261106</span>
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>Smallest<span class="w"> </span>scaffold:<span class="w"> </span><span class="m">7980</span>
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a><span class="c1"># contigs: 95</span>
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>Total<span class="w"> </span>contig<span class="w"> </span>length:<span class="w"> </span><span class="m">3024505387</span>
<a id="__codelineno-2-12" name="__codelineno-2-12" href="#__codelineno-2-12"></a>Average<span class="w"> </span>contig<span class="w"> </span>length:<span class="w"> </span><span class="m">31836898</span>.81
<a id="__codelineno-2-13" name="__codelineno-2-13" href="#__codelineno-2-13"></a>Contig<span class="w"> </span>N50:<span class="w"> </span><span class="m">101137168</span>
<a id="__codelineno-2-14" name="__codelineno-2-14" href="#__codelineno-2-14"></a>Contig<span class="w"> </span>auN:<span class="w"> </span><span class="m">98109753</span>.40
<a id="__codelineno-2-15" name="__codelineno-2-15" href="#__codelineno-2-15"></a>Contig<span class="w"> </span>L50:<span class="w"> </span><span class="m">12</span>
<a id="__codelineno-2-16" name="__codelineno-2-16" href="#__codelineno-2-16"></a>Largest<span class="w"> </span>contig:<span class="w"> </span><span class="m">201096255</span>
<a id="__codelineno-2-17" name="__codelineno-2-17" href="#__codelineno-2-17"></a>Smallest<span class="w"> </span>contig:<span class="w"> </span><span class="m">7980</span>
<a id="__codelineno-2-18" name="__codelineno-2-18" href="#__codelineno-2-18"></a><span class="c1"># gaps in scaffolds: 14</span>
<a id="__codelineno-2-19" name="__codelineno-2-19" href="#__codelineno-2-19"></a>Total<span class="w"> </span>gap<span class="w"> </span>length<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">1105310</span>
<a id="__codelineno-2-20" name="__codelineno-2-20" href="#__codelineno-2-20"></a>Average<span class="w"> </span>gap<span class="w"> </span>length<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">78950</span>.71
<a id="__codelineno-2-21" name="__codelineno-2-21" href="#__codelineno-2-21"></a>Gap<span class="w"> </span>N50<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">222184</span>
<a id="__codelineno-2-22" name="__codelineno-2-22" href="#__codelineno-2-22"></a>Gap<span class="w"> </span>auN<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">267123</span>.76
<a id="__codelineno-2-23" name="__codelineno-2-23" href="#__codelineno-2-23"></a>Gap<span class="w"> </span>L50<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">2</span>
<a id="__codelineno-2-24" name="__codelineno-2-24" href="#__codelineno-2-24"></a>Largest<span class="w"> </span>gap<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">425849</span>
<a id="__codelineno-2-25" name="__codelineno-2-25" href="#__codelineno-2-25"></a>Smallest<span class="w"> </span>gap<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">1000</span>
<a id="__codelineno-2-26" name="__codelineno-2-26" href="#__codelineno-2-26"></a>Base<span class="w"> </span>composition<span class="w"> </span><span class="o">(</span>A:C:G:T<span class="o">)</span>:<span class="w"> </span><span class="m">893171444</span>:618413603:615928136:896992204
<a id="__codelineno-2-27" name="__codelineno-2-27" href="#__codelineno-2-27"></a>GC<span class="w"> </span>content<span class="w"> </span>%:<span class="w"> </span><span class="m">40</span>.81
<a id="__codelineno-2-28" name="__codelineno-2-28" href="#__codelineno-2-28"></a><span class="c1"># soft-masked bases: 0</span>
<a id="__codelineno-2-29" name="__codelineno-2-29" href="#__codelineno-2-29"></a><span class="c1"># segments: 95</span>
<a id="__codelineno-2-30" name="__codelineno-2-30" href="#__codelineno-2-30"></a>Total<span class="w"> </span>segment<span class="w"> </span>length:<span class="w"> </span><span class="m">3024505387</span>
<a id="__codelineno-2-31" name="__codelineno-2-31" href="#__codelineno-2-31"></a>Average<span class="w"> </span>segment<span class="w"> </span>length:<span class="w"> </span><span class="m">31836898</span>.81
<a id="__codelineno-2-32" name="__codelineno-2-32" href="#__codelineno-2-32"></a><span class="c1"># gaps: 14</span>
<a id="__codelineno-2-33" name="__codelineno-2-33" href="#__codelineno-2-33"></a><span class="c1"># paths: 81</span>
</code></pre></div>
</details>
</div>
<h3 id="run-gfastats-on-a-gfa">Run gfastats on a GFA<a class="headerlink" href="#run-gfastats-on-a-gfa" title="Permanent link">&para;</a></h3>
<p>Remember that the file we initially got was an assembly <em>graph</em>&mdash;what if we wanted to know some graph-specific stats about our assembly, such as number of nodes or disconnected components? We can also assess that using gfastats. </p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/5-untip/unitig-normal-connected-tip.gfa<span class="w"> </span>.
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>gfastats<span class="w"> </span>--discover-paths<span class="w"> </span>unitig-normal-connected-tip.gfa
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>+++Assembly<span class="w"> </span>summary+++:<span class="w"> </span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="c1"># scaffolds: 2312</span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>Total<span class="w"> </span>scaffold<span class="w"> </span>length:<span class="w"> </span><span class="m">4179275425</span>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>Average<span class="w"> </span>scaffold<span class="w"> </span>length:<span class="w"> </span><span class="m">1807645</span>.08
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>Scaffold<span class="w"> </span>N50:<span class="w"> </span><span class="m">10829035</span>
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>Scaffold<span class="w"> </span>auN:<span class="w"> </span><span class="m">11407399</span>.64
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a>Scaffold<span class="w"> </span>L50:<span class="w"> </span><span class="m">134</span>
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>Largest<span class="w"> </span>scaffold:<span class="w"> </span><span class="m">38686347</span>
<a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a>Smallest<span class="w"> </span>scaffold:<span class="w"> </span><span class="m">1987</span>
<a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a><span class="c1"># contigs: 2312</span>
<a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>Total<span class="w"> </span>contig<span class="w"> </span>length:<span class="w"> </span><span class="m">4179275425</span>
<a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a>Average<span class="w"> </span>contig<span class="w"> </span>length:<span class="w"> </span><span class="m">1807645</span>.08
<a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a>Contig<span class="w"> </span>N50:<span class="w"> </span><span class="m">10829035</span>
<a id="__codelineno-4-14" name="__codelineno-4-14" href="#__codelineno-4-14"></a>Contig<span class="w"> </span>auN:<span class="w"> </span><span class="m">11407399</span>.64
<a id="__codelineno-4-15" name="__codelineno-4-15" href="#__codelineno-4-15"></a>Contig<span class="w"> </span>L50:<span class="w"> </span><span class="m">134</span>
<a id="__codelineno-4-16" name="__codelineno-4-16" href="#__codelineno-4-16"></a>Largest<span class="w"> </span>contig:<span class="w"> </span><span class="m">38686347</span>
<a id="__codelineno-4-17" name="__codelineno-4-17" href="#__codelineno-4-17"></a>Smallest<span class="w"> </span>contig:<span class="w"> </span><span class="m">1987</span>
<a id="__codelineno-4-18" name="__codelineno-4-18" href="#__codelineno-4-18"></a><span class="c1"># gaps in scaffolds: 0</span>
<a id="__codelineno-4-19" name="__codelineno-4-19" href="#__codelineno-4-19"></a>Total<span class="w"> </span>gap<span class="w"> </span>length<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">0</span>
<a id="__codelineno-4-20" name="__codelineno-4-20" href="#__codelineno-4-20"></a>Average<span class="w"> </span>gap<span class="w"> </span>length<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">0</span>.00
<a id="__codelineno-4-21" name="__codelineno-4-21" href="#__codelineno-4-21"></a>Gap<span class="w"> </span>N50<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">0</span>
<a id="__codelineno-4-22" name="__codelineno-4-22" href="#__codelineno-4-22"></a>Gap<span class="w"> </span>auN<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">0</span>.00
<a id="__codelineno-4-23" name="__codelineno-4-23" href="#__codelineno-4-23"></a>Gap<span class="w"> </span>L50<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">0</span>
<a id="__codelineno-4-24" name="__codelineno-4-24" href="#__codelineno-4-24"></a>Largest<span class="w"> </span>gap<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">0</span>
<a id="__codelineno-4-25" name="__codelineno-4-25" href="#__codelineno-4-25"></a>Smallest<span class="w"> </span>gap<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds:<span class="w"> </span><span class="m">0</span>
<a id="__codelineno-4-26" name="__codelineno-4-26" href="#__codelineno-4-26"></a>Base<span class="w"> </span>composition<span class="w"> </span><span class="o">(</span>A:C:G:T<span class="o">)</span>:<span class="w"> </span><span class="m">1180405407</span>:909914907:908088683:1180866428
<a id="__codelineno-4-27" name="__codelineno-4-27" href="#__codelineno-4-27"></a>GC<span class="w"> </span>content<span class="w"> </span>%:<span class="w"> </span><span class="m">43</span>.50
<a id="__codelineno-4-28" name="__codelineno-4-28" href="#__codelineno-4-28"></a><span class="c1"># soft-masked bases: 0</span>
<a id="__codelineno-4-29" name="__codelineno-4-29" href="#__codelineno-4-29"></a><span class="c1"># segments: 2312</span>
<a id="__codelineno-4-30" name="__codelineno-4-30" href="#__codelineno-4-30"></a>Total<span class="w"> </span>segment<span class="w"> </span>length:<span class="w"> </span><span class="m">4179275425</span>
<a id="__codelineno-4-31" name="__codelineno-4-31" href="#__codelineno-4-31"></a>Average<span class="w"> </span>segment<span class="w"> </span>length:<span class="w"> </span><span class="m">1807645</span>.08
<a id="__codelineno-4-32" name="__codelineno-4-32" href="#__codelineno-4-32"></a><span class="c1"># gaps: 0</span>
<a id="__codelineno-4-33" name="__codelineno-4-33" href="#__codelineno-4-33"></a><span class="c1"># paths: 2312</span>
<a id="__codelineno-4-34" name="__codelineno-4-34" href="#__codelineno-4-34"></a><span class="c1"># edges: 5790</span>
<a id="__codelineno-4-35" name="__codelineno-4-35" href="#__codelineno-4-35"></a>Average<span class="w"> </span>degree:<span class="w"> </span><span class="m">2</span>.50
<a id="__codelineno-4-36" name="__codelineno-4-36" href="#__codelineno-4-36"></a><span class="c1"># connected components: 47</span>
<a id="__codelineno-4-37" name="__codelineno-4-37" href="#__codelineno-4-37"></a>Largest<span class="w"> </span>connected<span class="w"> </span>component<span class="w"> </span>length:<span class="w"> </span><span class="m">530557693</span>
<a id="__codelineno-4-38" name="__codelineno-4-38" href="#__codelineno-4-38"></a><span class="c1"># dead ends: 562</span>
<a id="__codelineno-4-39" name="__codelineno-4-39" href="#__codelineno-4-39"></a><span class="c1"># disconnected components: 184</span>
<a id="__codelineno-4-40" name="__codelineno-4-40" href="#__codelineno-4-40"></a>Total<span class="w"> </span>length<span class="w"> </span>disconnected<span class="w"> </span>components:<span class="w"> </span><span class="m">60717682</span>
<a id="__codelineno-4-41" name="__codelineno-4-41" href="#__codelineno-4-41"></a><span class="c1"># separated components: 231</span>
<a id="__codelineno-4-42" name="__codelineno-4-42" href="#__codelineno-4-42"></a><span class="c1"># bubbles: 8</span>
<a id="__codelineno-4-43" name="__codelineno-4-43" href="#__codelineno-4-43"></a><span class="c1"># circular segments: 11</span>
</code></pre></div>
</details>
</div>
<details class="note">
<summary>What's the <code>--discover-paths</code> flag for?</summary>
<p>gfastats tries to clearly distinguish contigs from segments, so it will not pick up on contigs in a GFA without paths defined. To get the contig stats as well as graph stats from these GFAs, you'll need to add the <code>--discover-paths</code> flag. </p>
</details>
<p>Check out the graph-specific statistics at the end of the output. </p>
<h3 id="compare-two-graphs-stats">Compare two graphs' stats<a class="headerlink" href="#compare-two-graphs-stats" title="Permanent link">&para;</a></h3>
<p>Now that we know how to get the statistics for one assembly, let's get them for two so we can actually compare them. We already compared a Verkko HiFi-only and HiFi+ONT graph visually, so let's do it with assembly stats this time. We're going to use a one-liner to put the assembly stats side-by-side, because it can be kind of cumbersome to scroll up and down between two separate command line runs and their outputs.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>paste<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>&lt;<span class="o">(</span>gfastats<span class="w"> </span>-t<span class="w"> </span>--discover-paths<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/1-buildGraph/hifi-resolved.gfa<span class="o">)</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>&lt;<span class="o">(</span>gfastats<span class="w"> </span>-t<span class="w"> </span>--discover-paths<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/5-untip/unitig-normal-connected-tip.gfa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">2</span><span class="o">)</span>
</code></pre></div>
<div class="admonition info">
<ol>
<li><code>paste</code> is a command that pastes two files side by side</li>
<li>The <code>&lt;(COMMAND)</code> syntax is called process substitution, and it passes the output of the command(s) inside the parentheses to another command (here it is passing the <code>gfastats</code> output to <code>paste</code>), and can be useful when using a pipe (<code>|</code>) might not be possible</li>
<li>The <code>-t</code> flag in gfastats specifies that the output should be tab-delimited, which makes it more computer-parsable</li>
<li>The <code>cut</code> command in the substitution is just getting the actual statistics column from the gfastats output, because the first column is the name of the statistic</li>
</ol>
</div>
<p>Your output should look something like this:</p>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="c1"># contigs       68252   2312</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>Total<span class="w"> </span>contig<span class="w"> </span>length<span class="w">     </span><span class="m">3858601893</span><span class="w">      </span><span class="m">4179275425</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>Average<span class="w"> </span>contig<span class="w"> </span>length<span class="w">   </span><span class="m">56534</span>.63<span class="w">        </span><span class="m">1807645</span>.08
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>Contig<span class="w"> </span>N50<span class="w">      </span><span class="m">223090</span><span class="w">  </span><span class="m">10829035</span>
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>Contig<span class="w"> </span>auN<span class="w">      </span><span class="m">638121</span>.24<span class="w">       </span><span class="m">11407399</span>.64
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>Contig<span class="w"> </span>L50<span class="w">      </span><span class="m">4232</span><span class="w">    </span><span class="m">134</span>
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>Largest<span class="w"> </span>contig<span class="w">  </span><span class="m">23667669</span><span class="w">        </span><span class="m">38686347</span>
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>Smallest<span class="w"> </span>contig<span class="w"> </span><span class="m">1608</span><span class="w">    </span><span class="m">1987</span>
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a><span class="c1"># gaps in scaffolds     0       0</span>
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>Total<span class="w"> </span>gap<span class="w"> </span>length<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds<span class="w">   </span><span class="m">0</span><span class="w">       </span><span class="m">0</span>
<a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a>Average<span class="w"> </span>gap<span class="w"> </span>length<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds<span class="w"> </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00
<a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>Gap<span class="w"> </span>N50<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds<span class="w">    </span><span class="m">0</span><span class="w">       </span><span class="m">0</span>
<a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a>Gap<span class="w"> </span>auN<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00
<a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a>Gap<span class="w"> </span>L50<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds<span class="w">    </span><span class="m">0</span><span class="w">       </span><span class="m">0</span>
<a id="__codelineno-6-15" name="__codelineno-6-15" href="#__codelineno-6-15"></a>Largest<span class="w"> </span>gap<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds<span class="w">        </span><span class="m">0</span><span class="w">       </span><span class="m">0</span>
<a id="__codelineno-6-16" name="__codelineno-6-16" href="#__codelineno-6-16"></a>Smallest<span class="w"> </span>gap<span class="w"> </span><span class="k">in</span><span class="w"> </span>scaffolds<span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">0</span>
<a id="__codelineno-6-17" name="__codelineno-6-17" href="#__codelineno-6-17"></a>Base<span class="w"> </span>composition<span class="w"> </span><span class="o">(</span>A:C:G:T<span class="o">)</span><span class="w">      </span><span class="m">1089358958</span>:840831652:838233402:1090177881<span class="w">       </span><span class="m">1180405407</span>:909914907:908088683:1180866428
<a id="__codelineno-6-18" name="__codelineno-6-18" href="#__codelineno-6-18"></a>GC<span class="w"> </span>content<span class="w"> </span>%<span class="w">    </span><span class="m">43</span>.51<span class="w">   </span><span class="m">43</span>.50
<a id="__codelineno-6-19" name="__codelineno-6-19" href="#__codelineno-6-19"></a><span class="c1"># soft-masked bases     0       0</span>
<a id="__codelineno-6-20" name="__codelineno-6-20" href="#__codelineno-6-20"></a><span class="c1"># segments      68252   2312</span>
<a id="__codelineno-6-21" name="__codelineno-6-21" href="#__codelineno-6-21"></a>Total<span class="w"> </span>segment<span class="w"> </span>length<span class="w">    </span><span class="m">3858601893</span><span class="w">      </span><span class="m">4179275425</span>
<a id="__codelineno-6-22" name="__codelineno-6-22" href="#__codelineno-6-22"></a>Average<span class="w"> </span>segment<span class="w"> </span>length<span class="w">  </span><span class="m">56534</span>.63<span class="w">        </span><span class="m">1807645</span>.08
<a id="__codelineno-6-23" name="__codelineno-6-23" href="#__codelineno-6-23"></a><span class="c1"># gaps  0       0</span>
<a id="__codelineno-6-24" name="__codelineno-6-24" href="#__codelineno-6-24"></a><span class="c1"># paths 68252   2312</span>
<a id="__codelineno-6-25" name="__codelineno-6-25" href="#__codelineno-6-25"></a><span class="c1"># edges 181274  5790</span>
<a id="__codelineno-6-26" name="__codelineno-6-26" href="#__codelineno-6-26"></a>Average<span class="w"> </span>degree<span class="w">  </span><span class="m">2</span>.66<span class="w">    </span><span class="m">2</span>.50
<a id="__codelineno-6-27" name="__codelineno-6-27" href="#__codelineno-6-27"></a><span class="c1"># connected components  45      47</span>
<a id="__codelineno-6-28" name="__codelineno-6-28" href="#__codelineno-6-28"></a>Largest<span class="w"> </span>connected<span class="w"> </span>component<span class="w"> </span>length<span class="w">      </span><span class="m">496518810</span><span class="w">       </span><span class="m">530557693</span>
<a id="__codelineno-6-29" name="__codelineno-6-29" href="#__codelineno-6-29"></a><span class="c1"># dead ends     684     562</span>
<a id="__codelineno-6-30" name="__codelineno-6-30" href="#__codelineno-6-30"></a><span class="c1"># disconnected components       18      184</span>
<a id="__codelineno-6-31" name="__codelineno-6-31" href="#__codelineno-6-31"></a>Total<span class="w"> </span>length<span class="w"> </span>disconnected<span class="w"> </span>components<span class="w">    </span><span class="m">7267912</span><span class="w"> </span><span class="m">60717682</span>
<a id="__codelineno-6-32" name="__codelineno-6-32" href="#__codelineno-6-32"></a><span class="c1"># separated components  63      231</span>
<a id="__codelineno-6-33" name="__codelineno-6-33" href="#__codelineno-6-33"></a><span class="c1"># bubbles       4312    8</span>
<a id="__codelineno-6-34" name="__codelineno-6-34" href="#__codelineno-6-34"></a><span class="c1"># circular segments     31      11</span>
</code></pre></div>
</details>
</div>
<div class="admonition info">
<p class="admonition-title">Output explained</p>
<p>... where the first column is the stats from the HiFi-only assembly graph, and the second column is the stats from the HiFi+ONT assembly graph. Notice how the HiFi-only graph has way more nodes than the HiFi+ONT one, like we'd seen in Bandage. Stats-wise, this results in the HiFi-only graph having a N50 value of 223 Kbp while the HiFi+ONT one is 10.8 Mbp, a whole order of magnitude larger. For the HiFi-only graph, though, there's a bigger difference between its N50 value and its auN value: 223 Kbp vs. 638 Kbp, while the HiFi+ONT stats have a smaller difference of 10.8 Mbp vs. 11.4 Mbp. This might be due to the HiFi-only graph having on average shorter segments and more of the shorter ones, since it doesn't have the ONT data to resolve the segments into larger ones. </p>
</div>
<h2 id="correctness-qv-using-merqury">Correctness (QV using Merqury)<a class="headerlink" href="#correctness-qv-using-merqury" title="Permanent link">&para;</a></h2>
<p>Correctness refers to the base pair accuracy, and can be measured by comparing one's assembly to a gold standard reference genome. This approach is limited by 1) an assumption about the quality of the reference itself and the closeness between it and the assembly being compared, and 2) the need for a reference genome at all, which many species do not have (yet). To avoid this, we can use <strong>Merqury</strong>: a reference-free suite of tools for assessing assembly quality (particularly w.r.t. error rate) using <em>k</em>-mers and the read set that generated that assembly. If an assembly is made up from the same sequences that were in the sequencing reads, then we would not expect any sequences (<em>k</em>-mers) in the assembly that aren't present in the read set&mdash;but we do find those sometimes, and those are what Merqury flags as error <em>k</em>-mers. <strong>It uses the following formula to calculate QV value, which typically results in QVs of 50-60</strong>: </p>
<div class="arithmatex">\[
\Large 
    E=1-P=1-\left (1-\frac{K_{\textrm{asm}}}{K_{\textrm{total}}} \right ) ^{\frac{1}{k}}
\]</div>
<!--
<center>
![QV formula](https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/merqury_qvformula.png)
</center>
-->

<details class="note">
<summary>OK, but what does 'QV' mean, anyway?</summary>
<p>The QV that Merqury is interpreted similarly to the commonly used Phred quality scale, which might be familiar to those who have done short-read sequencing or are otherwise acquainted with FASTQ files. Phred quality scores are logarithmically related to error-probability, such that:
- Phred score of 30 represents a 1 in 1,000 error probability (<em>i.e.</em>, 99.9% accuracy)
- Phred score of 40 represents a 1 in 10,000 error probability (<em>i.e.</em>, 99.99% accuracy)
- Phred score of 50 represents a 1 in 100,000 error probability (<em>i.e.</em>, 99.999% accuracy)
- Phred score of 60 represents a 1 in 1,000,000 error probability (<em>i.e.</em>, 99.9999% accuracy)</p>
</details>
<p>Merqury operates using <em>k</em>-mer databases like the ones we generated using meryl, so that's what we'll do now. </p>
<h3 id="running-meryl-and-genomescope-on-the-e-coli-verkko-assembly">Running Meryl and GenomeScope on the <em>E. coli</em> Verkko assembly<a class="headerlink" href="#running-meryl-and-genomescope-on-the-e-coli-verkko-assembly" title="Permanent link">&para;</a></h3>
<p>Let's try this out on the <em>E. coli</em> Verkko assembly. First we need a Meryl database, so let's generate that.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="nb">cd</span><span class="w"> </span>~/lra
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>day3_assembly_qc/merqury
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="nb">cd</span><span class="w"> </span>day3_assembly_qc/merqury
</code></pre></div>
We just made a directory for our runs, now let's sym link the fasta and reads here so we can refer to them more easily
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>ln -s ../../day2_assembly/verkko_test/assembly/assembly.fasta .
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>ln -s ../../day2_assembly/verkko_test/hifi.fastq.gz .
</code></pre></div>
Now we can run it!
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>module purge
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>module load Merqury
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>sbatch --account nesi02659 --job-name=meryl --time=00:15:00 --cpus-per-task 8 --mem=24G -p milan --wrap=&quot;meryl count k=30 memory=24 threads=8 hifi.fastq.gz output read-db.meryl&quot;
</code></pre></div></p>
</div>
<details class="note">
<summary><code>--wrap</code> ???</summary>
<p>Previously, we used the <code>sbatch</code> command to submit a slurm script to the cluster and the slurm job handler. The <code>sbatch</code> command can actually take a lot of parameters like the ones we included in the beginning of our script, and one of those parameters is <code>--wrap</code> which kind of wraps whatever command you give it in a Slurm wrapper so that the cluster can schedule it as if it was a Slurm script. </p>
<p>Take note that running a process in this manner is not reproducible. Unless you have access to the <code>history</code> logs, other researchers are not able to know what parameters you have used. Therefore, it is advisable to write it out in a Slurm script as below:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="c1">#SBATCH --account=nesi02659</span>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="c1">#SBATCH --job-name=meryl</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="c1">#SBATCH --time=00:15:00</span>
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a><span class="c1">#SBATCH --cpus-per-task=8</span>
<a id="__codelineno-10-6" name="__codelineno-10-6" href="#__codelineno-10-6"></a><span class="c1">#SBATCH --mem=24G</span>
<a id="__codelineno-10-7" name="__codelineno-10-7" href="#__codelineno-10-7"></a><span class="c1">#SBATCH --partition=milan</span>
<a id="__codelineno-10-8" name="__codelineno-10-8" href="#__codelineno-10-8"></a>
<a id="__codelineno-10-9" name="__codelineno-10-9" href="#__codelineno-10-9"></a><span class="c1"># Modules</span>
<a id="__codelineno-10-10" name="__codelineno-10-10" href="#__codelineno-10-10"></a>module<span class="w"> </span>purge
<a id="__codelineno-10-11" name="__codelineno-10-11" href="#__codelineno-10-11"></a>module<span class="w"> </span>load<span class="w"> </span>Merqury/1.3-Miniconda3
<a id="__codelineno-10-12" name="__codelineno-10-12" href="#__codelineno-10-12"></a>
<a id="__codelineno-10-13" name="__codelineno-10-13" href="#__codelineno-10-13"></a><span class="c1"># Run</span>
<a id="__codelineno-10-14" name="__codelineno-10-14" href="#__codelineno-10-14"></a>meryl<span class="w"> </span>count<span class="w"> </span><span class="nv">k</span><span class="o">=</span><span class="m">30</span><span class="w"> </span><span class="nv">memory</span><span class="o">=</span><span class="m">24</span><span class="w"> </span><span class="nv">threads</span><span class="o">=</span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-15" name="__codelineno-10-15" href="#__codelineno-10-15"></a><span class="w">  </span>hifi.fastq.gz<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-16" name="__codelineno-10-16" href="#__codelineno-10-16"></a><span class="w">  </span>output<span class="w"> </span>read-db.meryl
</code></pre></div>
</div>
</details>
<p>That shouldn't take too long to run. Now we have a Meryl DB for our HiFi reads. If we're curious about the distribution of our <em>k</em>-mers, we can use Meryl generate a histogram of the counts to show us how often a <em>k</em>-mer occurs only once in the reads, twice, etc. </p>
<details class="question">
<summary>How would you go about trying to do this with meryl?</summary>
<p>When you want to use a tool to do something (and you are decently confident that the tool can actually do it), then a good point to start is just querying the tool's manual or help dialogue. Try out <code>meryl --help</code> and see if there's a function that looks like it could generate the histogram we want. <del>spoiler alert: it's <code>meryl histogram read-db.meryl</code></del></p>
</details>
<p>If you tried to run that command with the output straight to standard out (<em>i.e.</em>, your terminal screen), you'll see it's rather overwhelming and puts you all the way at the high, high coverage <em>k</em>-mer counts, which are only one occurrence. Let's look at just the first 100 lines instead.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>meryl<span class="w"> </span>histogram<span class="w"> </span>read-db.meryl<span class="w"> </span>&gt;<span class="w"> </span>read-db.hist
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>head<span class="w"> </span>-n<span class="w"> </span><span class="m">100</span><span class="w"> </span>read-db.hist
</code></pre></div>
</div>
<p>This is more manageable, and you can even kind of see the histogram forming from the count values. There's a lot of <em>k</em>-mers that are present at only one copy (or otherwise very low copy) in the read set: these are usually sequencing errors, because there's a lot of these <em>k</em>-mers present at low copy. Because the sequence isn't actually real (<em>i.e.</em>, it isn't actually in the genome and isn't actually serving as sequencing template), these <em>k</em>-mers stay at low copy. After these error <em>k</em>-mers, there's a dip in the histogram until about the 24&ndash;28 copy range. This peak is the coverage of the actual <em>k</em>-mers coming from the genome that you sequenced, thus it corresponds to having coverage of ~26X in this read set. We only have one peak here because this is a haploid dataset, but if your dataset is diploid then expect two peaks with the first peak varying in height depending on heterozygosity of your sample. </p>
<p>"What if I want a pretty graph instead of imagining it?" Good news&mdash;there's <del>an app</del> a program for that. GenomeScope is a straightforward program with an online web page where you can just drop in your meryl histogram file and it will draw the histogram for you as well as use the GenomeScope model to predict some genome characteristics of your data, given the expected ploidy. Let's try it out! Download the <code>read-db.hist</code> file and throw it into the GenomeScope website: <a href="http://qb.cshl.edu/genomescope/genomescope2.0/">http://qb.cshl.edu/genomescope/genomescope2.0/</a> and adjust the parameters accordingly.</p>
<details class="tip">
<summary>Can I use GenomeScope to QC my raw data before assembly?</summary>
<p>As you can see here, GenomeScope can be useful for getting an idea of what your raw dataset looks like, as well as feeling out the genome that should be represented by those sequencing reads. This can be useful as a QC step before even running the assembly, to make sure that your dataset is good enough to use. Here's an example of a good diploid dataset of HiFi reads:</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/genomescope_mMicPen1.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="mMicPen1 genomescope" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/genomescope_mMicPen1.png" /></a></p>
<details class="question">
<summary>How does the data look? What does the coverage look to be? How many peaks are there in the data and what do they represent? What are some characteristics of the genome as inferred by GenomeScope?</summary>
<p>This data looks good, and you know that 1) because this tutorial text already called it good previously, and 2) there's a good amount of coverage, around 40X diploid coverage in fact. Additionally, the peaks are all very clear and distinct from each other and from the error <em>k</em>-mer slope on the left. Recall that the first peak represents haploid coverage (i.e., coverage of heterozygous loci) and the second peak is diploid coverage. GenomeScope is predicting the total size of the genome to be about 2.2 Gbp with 1.23% heterozygosity. This is data for <i>Microtus pennsylvaticus</i>, the eastern meadow vole. </p>
</details>
<p>Here's an example of another HiFi dataset:</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/genomescope_xbAnaTube1_bad.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="xbAnaTube1 genomescope 1" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/genomescope_xbAnaTube1_bad.png" /></a></p>
<p>Compare this to the previous examples. Does this look like a good dataset? Ignoring how GenomeScope itself is getting confused and can't get its model to fit properly to the <em>k</em>-mer spectra, let's look at the actual observed <em>k</em>-mer spectra. It does look like there's potentially two peaks whose distributions are overlapping, one peak around 10X and the other just under 20X. These are presumably our haploid and diploid peaks, but there's not enough coverage to resolve them properly here. This is an example of a GenomeScope QC that would tell me we don't have enough HiFi data to continue onto assembly, let's try to generate some more data.  </p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/genomescope_xbAnaTube1_good.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="xbAnaTube1 genomescope 2" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/genomescope_xbAnaTube1_good.png" /></a></p>
<p>Wow, more data! This result is from after adding one more (notably more successful) SMRT cell of HiFi data and re-running GenomeScope. We can see the resolution of the peaks much more cleanly, and the GenomeScope model fits the data much better now, so we can trust the genome characteristic estimates here more than before. If you noted the comparatively small genome size and wondered what this was, it's <em>Anadara tuberculosa</em>, the piangua.</p>
<p>Now you might be wondering: what happens if I try to assemble data without enough coverage? Answer: a headache. The assembly that results from the dataset that made the first GenomeScope plot resulted in two haplotypes of over 3,000 contigs each, which is very fragmented for a genome this small, recapitulated by their auN values being ~0.5 Mbp. In comparison, an assembly with the dataset pictured in the second GenomeScope plot resulted in two haplotypes of 500-800 contigs with auN values of 3.6-3.9 Mbp! The improvement in contiguity can also be visualized in the Bandage plots:</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/bandage_xbAnaTube1_bad.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="xbAnaTube1 bandage 1" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/bandage_xbAnaTube1_bad.png" /></a></p>
<p>The above is the Hifiasm unitig graph for the assembly done without good HiFi coverage.</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/bandage_xbAnaTube1_good.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="xbAnaTube1 bandage 2" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/bandage_xbAnaTube1_good.png" /></a></p>
<p>The above is the Hifiasm unitig graph for the assembly done with good (~56X) HiFi coverage.</p>
</details>
<h3 id="using-merqury-in-solo-mode">Using Merqury in solo mode<a class="headerlink" href="#using-merqury-in-solo-mode" title="Permanent link">&para;</a></h3>
<p>Let's use Merqury on that database we just made to get the QV and some plots.</p>
<p>Use your text editor of choice to make a Slurm script (<code>run_merqury.sl</code>) to run the actual Merqury program with the following contents:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a><span class="c1">#SBATCH --job-name      merqury1</span>
<a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a><span class="c1">#SBATCH --cpus-per-task 8</span>
<a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a><span class="c1">#SBATCH --time          00:15:00</span>
<a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a><span class="c1">#SBATCH --mem           40G</span>
<a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a><span class="c1">#SBATCH --output        slurmlogs/%x.%j.log</span>
<a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a><span class="c1">#SBATCH --error         slurmlogs/%x.%j.err</span>
<a id="__codelineno-12-9" name="__codelineno-12-9" href="#__codelineno-12-9"></a><span class="c1">#SBATCH --partition     milan</span>
<a id="__codelineno-12-10" name="__codelineno-12-10" href="#__codelineno-12-10"></a>
<a id="__codelineno-12-11" name="__codelineno-12-11" href="#__codelineno-12-11"></a><span class="c1">## load modules</span>
<a id="__codelineno-12-12" name="__codelineno-12-12" href="#__codelineno-12-12"></a>module<span class="w"> </span>purge
<a id="__codelineno-12-13" name="__codelineno-12-13" href="#__codelineno-12-13"></a>module<span class="w"> </span>load<span class="w"> </span>Merqury
<a id="__codelineno-12-14" name="__codelineno-12-14" href="#__codelineno-12-14"></a><span class="nb">export</span><span class="w"> </span><span class="nv">MERQURY</span><span class="o">=</span>/opt/nesi/CS400_centos7_bdw/Merqury/1.3-Miniconda3/merqury
<a id="__codelineno-12-15" name="__codelineno-12-15" href="#__codelineno-12-15"></a>
<a id="__codelineno-12-16" name="__codelineno-12-16" href="#__codelineno-12-16"></a><span class="c1">## create solo merqury dir and use it</span>
<a id="__codelineno-12-17" name="__codelineno-12-17" href="#__codelineno-12-17"></a>mkdir<span class="w"> </span>merqury_solo
<a id="__codelineno-12-18" name="__codelineno-12-18" href="#__codelineno-12-18"></a><span class="nb">cd</span><span class="w"> </span>merqury_solo
<a id="__codelineno-12-19" name="__codelineno-12-19" href="#__codelineno-12-19"></a>
<a id="__codelineno-12-20" name="__codelineno-12-20" href="#__codelineno-12-20"></a><span class="c1">## run merqury</span>
<a id="__codelineno-12-21" name="__codelineno-12-21" href="#__codelineno-12-21"></a>merqury.sh<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-22" name="__codelineno-12-22" href="#__codelineno-12-22"></a><span class="w">    </span>../read-db.meryl<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-23" name="__codelineno-12-23" href="#__codelineno-12-23"></a><span class="w">    </span>../assembly.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-24" name="__codelineno-12-24" href="#__codelineno-12-24"></a><span class="w">    </span>output
<a id="__codelineno-12-25" name="__codelineno-12-25" href="#__codelineno-12-25"></a>
<a id="__codelineno-12-26" name="__codelineno-12-26" href="#__codelineno-12-26"></a><span class="nb">cd</span><span class="w"> </span>-/lra
</code></pre></div>
</div>
<details class="note">
<summary>What's that <code>export</code> command doing there?</summary>
<p>Merqury as a package ships with a lot of scripts, especially for plotting. The <code>merqury.sh</code> command that we're using is calling those scripts, but we need to tell it where we installed Merqury. </p>
</details>
<p>To find out the QV, we want the file named <code>output.qv</code>. Take a look at it and try to interpret the QV value you find (third column). If we recall the Phred scale system, this would mean that this QV value is great! Which is not surprising, considering we used HiFi data. <strong>It's worth noting, though, that we are using HiFi <em>k</em>-mers to evaluate sequences derived from those same HiFi reads.</strong> This does a good job of showing whether the assembly worked with that data well, but what if the HiFi data itself is missing parts of the genome, such as due to bias (<em>e.g.</em>, GA dropout)? That's why it's important to use orthogonal datasets made using different sequencing technology, when possible. For instance, we can use an Illumina-based Meryl database to evaluate a HiFi assembly. For non-human vertebrates, this often results in the QV dropping from 50-60 to 35-45, depending on the genome in question. </p>
<h3 id="using-merqury-in-trio-mode">Using Merqury in trio mode<a class="headerlink" href="#using-merqury-in-trio-mode" title="Permanent link">&para;</a></h3>
<p>So we just ran Merqury on our E. coli assembly, and evaluated it using the HiFi reads that we used for that assembly. Merqury can also utilize trio data (using those hapmer DBs we talked about before) to evaluate the phasing of an assembly, so let's try that with our HG002 trio data. </p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="c1">#SBATCH --job-name      merqury2</span>
<a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="c1">#SBATCH --cpus-per-task 8</span>
<a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="c1">#SBATCH --time          02:00:00</span>
<a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a><span class="c1">#SBATCH --mem           40G</span>
<a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="c1">#SBATCH --output        slurmlogs/%x.%j.log</span>
<a id="__codelineno-13-8" name="__codelineno-13-8" href="#__codelineno-13-8"></a><span class="c1">#SBATCH --error         slurmlogs/%x.%j.err</span>
<a id="__codelineno-13-9" name="__codelineno-13-9" href="#__codelineno-13-9"></a><span class="c1">#SBATCH --partition     milan</span>
<a id="__codelineno-13-10" name="__codelineno-13-10" href="#__codelineno-13-10"></a>
<a id="__codelineno-13-11" name="__codelineno-13-11" href="#__codelineno-13-11"></a><span class="c1">## load modules</span>
<a id="__codelineno-13-12" name="__codelineno-13-12" href="#__codelineno-13-12"></a>module<span class="w"> </span>purge
<a id="__codelineno-13-13" name="__codelineno-13-13" href="#__codelineno-13-13"></a>module<span class="w"> </span>load<span class="w"> </span>Merqury
<a id="__codelineno-13-14" name="__codelineno-13-14" href="#__codelineno-13-14"></a><span class="nb">export</span><span class="w"> </span><span class="nv">MERQURY</span><span class="o">=</span>/opt/nesi/CS400_centos7_bdw/Merqury/1.3-Miniconda3/merqury
<a id="__codelineno-13-15" name="__codelineno-13-15" href="#__codelineno-13-15"></a>
<a id="__codelineno-13-16" name="__codelineno-13-16" href="#__codelineno-13-16"></a><span class="c1">## create trio merqury dir and use it</span>
<a id="__codelineno-13-17" name="__codelineno-13-17" href="#__codelineno-13-17"></a><span class="nb">cd</span><span class="w"> </span>~/lra/day3_assembly_qc
<a id="__codelineno-13-18" name="__codelineno-13-18" href="#__codelineno-13-18"></a>mkdir<span class="w"> </span>merqury_trio
<a id="__codelineno-13-19" name="__codelineno-13-19" href="#__codelineno-13-19"></a><span class="nb">cd</span><span class="w"> </span>merqury_trio
<a id="__codelineno-13-20" name="__codelineno-13-20" href="#__codelineno-13-20"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/assembly.*.fasta<span class="w"> </span>.
<a id="__codelineno-13-21" name="__codelineno-13-21" href="#__codelineno-13-21"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/maternal.k30.hapmer.meryl<span class="w"> </span>.
<a id="__codelineno-13-22" name="__codelineno-13-22" href="#__codelineno-13-22"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/paternal.k30.hapmer.meryl<span class="w"> </span>.
<a id="__codelineno-13-23" name="__codelineno-13-23" href="#__codelineno-13-23"></a>
<a id="__codelineno-13-24" name="__codelineno-13-24" href="#__codelineno-13-24"></a><span class="c1">## let&#39;s run the program in a results directory to make things a little neater</span>
<a id="__codelineno-13-25" name="__codelineno-13-25" href="#__codelineno-13-25"></a>mkdir<span class="w"> </span>results
<a id="__codelineno-13-26" name="__codelineno-13-26" href="#__codelineno-13-26"></a><span class="nb">cd</span><span class="w"> </span>results
<a id="__codelineno-13-27" name="__codelineno-13-27" href="#__codelineno-13-27"></a>
<a id="__codelineno-13-28" name="__codelineno-13-28" href="#__codelineno-13-28"></a><span class="c1">## run merqury</span>
<a id="__codelineno-13-29" name="__codelineno-13-29" href="#__codelineno-13-29"></a>merqury.sh<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-30" name="__codelineno-13-30" href="#__codelineno-13-30"></a><span class="w">    </span>../read-db.meryl<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-31" name="__codelineno-13-31" href="#__codelineno-13-31"></a><span class="w">    </span>../paternal.k30.hapmer.meryl<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-32" name="__codelineno-13-32" href="#__codelineno-13-32"></a><span class="w">    </span>../maternal.k30.hapmer.meryl<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-33" name="__codelineno-13-33" href="#__codelineno-13-33"></a><span class="w">    </span>../assembly.haplotype1.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-34" name="__codelineno-13-34" href="#__codelineno-13-34"></a><span class="w">    </span>../assembly.haplotype2.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-35" name="__codelineno-13-35" href="#__codelineno-13-35"></a><span class="w">    </span>output
<a id="__codelineno-13-36" name="__codelineno-13-36" href="#__codelineno-13-36"></a>
<a id="__codelineno-13-37" name="__codelineno-13-37" href="#__codelineno-13-37"></a><span class="nb">cd</span><span class="w"> </span>-/lra
</code></pre></div>
</div>
<p>You can also look in this folder for the merqury outputs if there isn't enough time to run the actual program: <code>/nesi/nobackup/nesi02659/LRA/resources/merqury</code></p>
<p>There will be lots of outputs, but let's look at two now to get an idea for how the sequences have been partitioned between our assemblies, and whether that's consistent with information we know from trio sequencing. First let's look at the spectra-asm plot:</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/GenomicsAotearoa/long-read-assembly/main/docs/images/qc/verkkotrio.spectra-asm.fl.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="verkko trio spectra" src="https://raw.githubusercontent.com/GenomicsAotearoa/long-read-assembly/main/docs/images/qc/verkkotrio.spectra-asm.fl.png" /></a></p>
<p>Merqury's spectra plots take the kmer spectrum (like you'd seen in GenomeScope) and color the kmers according to where the kmer is found: the reads <em>only</em>, one of the assemblies, or both of the assemblies.</p>
<p>A useful output from Merqury for evaluating phasing is the blob plot:</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/GenomicsAotearoa/long-read-assembly/main/docs/images/qc/verkkotrio.hapmers.blob.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="verkko trio blob" src="https://raw.githubusercontent.com/GenomicsAotearoa/long-read-assembly/main/docs/images/qc/verkkotrio.hapmers.blob.png" /></a></p>
<p>In this plot, each blob is a contig, and its x,y position represents parental hapmer content, while color represents assembly-of-origin.</p>
<h3 id="what-do-different-phasing-approaches-look-like-in-merqury">What do different phasing approaches look like in Merqury?<a class="headerlink" href="#what-do-different-phasing-approaches-look-like-in-merqury" title="Permanent link">&para;</a></h3>
<p>Now we know what a nice trio-phased assembly looks like in Merqury, but what do the other options (Hi-C phasing or no phasing at all) look like? Let's look at an example from the zebra finch (<em>Taeniopygia guttata</em>), where the Vertebrate Genomes Project (VGP) has used hifiasm on the same HiFi dataset with different phasing approaches and evaluated the resulting assemblies with trio data in order to benchmark the different methods. </p>
<p><a class="glightbox" href="https://github.com/GenomicsAotearoa/long-read-assembly/blob/main/docs/images/qc/merqury_bTaeGut2_blobs.png?raw=true" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="bTaeGut2 examples" src="https://github.com/GenomicsAotearoa/long-read-assembly/blob/main/docs/images/qc/merqury_bTaeGut2_blobs.png?raw=true" /></a></p>
<p>On the left, we have the blob plot for the trio assembly, which looks nicely phased as we expect -- all the contigs exhibit hapmers from only one parent (which you can tell because all blobs are along the x- or y- axis), and on top of that all the contigs from one assembly show only one parent's hapmers (<em>i.e.</em>, hap1's red blobs all show only bTaeGut3 hapmers). In the middle, we have a blob plot for a primary/alternate set of assemblies generated without any phasing data. You'll notice that a lot of the primary assembly blobs are not on the axes, meaning they have hapmers from both parents. </p>
<details class="question">
<summary>Why are the alternate contigs all in phase?</summary>
<p>It starts to make sense if you think about what the alternae assembly is meant to represent: the haplotigs. These are the alternate alleles for the heterozygous loci, so it would track that only one parent's hapmers are represented -- the alternate assembly is a collection of the "other side of the bubble"s when looking at the assembly graph. </p>
</details>
<p>The right plot is a blob plot from the Hi-C-phased assembly. Notably, most of the contigs are able to be properly phased, since they are almost all on the x- or -y axis. </p>
<h3 id="switch-and-hamming-errors-using-yak">Switch and Hamming errors using yak<a class="headerlink" href="#switch-and-hamming-errors-using-yak" title="Permanent link">&para;</a></h3>
<p>Two more types of errors we use to assess assemblies are switch errors and Hamming errors. Hamming errors represent the percentage of SNPs wrongly phased (compared to ground truth), while switch errors represent the percentage of <em>adjacent</em> SNP pairs wrongly phased. See the following graphic:</p>
<p><a class="glightbox" href="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/yak_switcherror.png" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="switch errors" src="https://raw.githubusercontent.com/human-pangenomics/hprc-tutorials/GA-workshop/assembly/genomics_aotearoa/images/qc/yak_switcherror.png" /></a></p>
<p>As the image illustrates, switch errors occur when an assembly <em>switches</em> between haplotypes. These errors are more prevalent in pseudohaplotype (<em>e.g.</em>, primary/alternate) assemblies that did not use any phasing data, as the assembler has no way of properly linking haplotype blocks, which can result in mixed hapmer content contigs that are a chimera of parental sequences. </p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a><span class="c1">#SBATCH --job-name      yaktrioeval</span>
<a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a><span class="c1">#SBATCH --cpus-per-task 32</span>
<a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a><span class="c1">#SBATCH --time          01:00:00</span>
<a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a><span class="c1">#SBATCH --mem           256G</span>
<a id="__codelineno-14-7" name="__codelineno-14-7" href="#__codelineno-14-7"></a><span class="c1">#SBATCH --output        slurmlogs/%x.%j.log</span>
<a id="__codelineno-14-8" name="__codelineno-14-8" href="#__codelineno-14-8"></a><span class="c1">#SBATCH --error         slurmlogs/%x.%j.err</span>
<a id="__codelineno-14-9" name="__codelineno-14-9" href="#__codelineno-14-9"></a><span class="c1">#SBATCH --partition     milan</span>
<a id="__codelineno-14-10" name="__codelineno-14-10" href="#__codelineno-14-10"></a>
<a id="__codelineno-14-11" name="__codelineno-14-11" href="#__codelineno-14-11"></a><span class="c1">## change to qc dir, link things</span>
<a id="__codelineno-14-12" name="__codelineno-14-12" href="#__codelineno-14-12"></a><span class="nb">cd</span><span class="w"> </span>~/lra/day3_assembly_qc
<a id="__codelineno-14-13" name="__codelineno-14-13" href="#__codelineno-14-13"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/yak<span class="w"> </span>.
<a id="__codelineno-14-14" name="__codelineno-14-14" href="#__codelineno-14-14"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/hifiasm/full/hic<span class="w"> </span>.
<a id="__codelineno-14-15" name="__codelineno-14-15" href="#__codelineno-14-15"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/hifiasm/full/trio<span class="w"> </span>.
<a id="__codelineno-14-16" name="__codelineno-14-16" href="#__codelineno-14-16"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>qc_yak
<a id="__codelineno-14-17" name="__codelineno-14-17" href="#__codelineno-14-17"></a><span class="nb">cd</span><span class="w"> </span>qc_yak
<a id="__codelineno-14-18" name="__codelineno-14-18" href="#__codelineno-14-18"></a>
<a id="__codelineno-14-19" name="__codelineno-14-19" href="#__codelineno-14-19"></a><span class="c1">## load modules</span>
<a id="__codelineno-14-20" name="__codelineno-14-20" href="#__codelineno-14-20"></a>module<span class="w"> </span>purge
<a id="__codelineno-14-21" name="__codelineno-14-21" href="#__codelineno-14-21"></a>module<span class="w"> </span>load<span class="w"> </span>yak
<a id="__codelineno-14-22" name="__codelineno-14-22" href="#__codelineno-14-22"></a>
<a id="__codelineno-14-23" name="__codelineno-14-23" href="#__codelineno-14-23"></a><span class="c1">## run yak</span>
<a id="__codelineno-14-24" name="__codelineno-14-24" href="#__codelineno-14-24"></a>yak<span class="w"> </span>trioeval<span class="w"> </span>-t<span class="w"> </span><span class="m">32</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-25" name="__codelineno-14-25" href="#__codelineno-14-25"></a><span class="w">    </span>../yak/pat.HG003.yak<span class="w"> </span>../yak/mat.HG004.yak<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-26" name="__codelineno-14-26" href="#__codelineno-14-26"></a><span class="w">    </span>../hic/HG002.hap1.fa.gz<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-27" name="__codelineno-14-27" href="#__codelineno-14-27"></a><span class="w">    </span>&gt;<span class="w"> </span>hifiasm.hic.hap1.trioeval
<a id="__codelineno-14-28" name="__codelineno-14-28" href="#__codelineno-14-28"></a>
<a id="__codelineno-14-29" name="__codelineno-14-29" href="#__codelineno-14-29"></a>yak<span class="w"> </span>trioeval<span class="w"> </span>-t<span class="w"> </span><span class="m">32</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-30" name="__codelineno-14-30" href="#__codelineno-14-30"></a><span class="w">    </span>../yak/pat.HG003.yak<span class="w"> </span>../yak/mat.HG004.yak<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-31" name="__codelineno-14-31" href="#__codelineno-14-31"></a><span class="w">    </span>../trio/HG002.mat.fa.gz<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-32" name="__codelineno-14-32" href="#__codelineno-14-32"></a><span class="w">    </span>&gt;<span class="w"> </span>hifiasm.trio.mat.trioeval
</code></pre></div>
</div>
<h2 id="completeness-asmgene">Completeness (asmgene)<a class="headerlink" href="#completeness-asmgene" title="Permanent link">&para;</a></h2>
<p>Another way to assess an assembly is via <strong>completeness</strong>, particularly with regard to expected gene content. If you have a reference genome that's been annotated with coding sequences, then you can use the tool <em>asmgene</em> to align multi-copy genes to your assembly and see if they remain multi-copy, or if the assembler has created a misassembly. asmgene works by aligning annotated transcripts to the reference genome, and record hits if the transcript is mapped at or over 99% identity over 99% or greater of the transcript length. If the transcript only has one hit, then it is single-copy (SC), otherwise it's multi-copy (MC). The same is then done for your assembly, and the fraction of missing multi-copy (%MMC) gene content is computed. </p>
<p><a class="glightbox" href="https://github.com/GenomicsAotearoa/long-read-assembly/blob/main/docs/images/qc/asmgene.png?raw=true" data-type="image" data-width="100%" data-height="auto" data-desc-position="bottom"><img alt="asmgene" src="https://github.com/GenomicsAotearoa/long-read-assembly/blob/main/docs/images/qc/asmgene.png?raw=true" /></a></p>
<p>A perfect asesmbly would have %MMC be zero, while a higher fraction indicates the assembly has collapsed some of these multi-copy genes. Additionally, you can look at the presence (or absence!) of expected single-copy genes in order to check gene completeness of the assembly. </p>
<p>The output will be a tab-delimed list of metrics and the value of that metric for the reference and for your given assembly. The line <strong>full_sgl</strong> gives the number of single-copy genes present in the reference and your assembly&mdash;if these numbers are off-balanced, then you might have false duplications, which are also pointed out on the <strong>full_dup</strong> line. For the multi-copy genes, you can look at <strong>dup_cnt</strong> to see the number of multi-copy genes in the reference and see how many of those genes are still multi-copy in your assembly. You can then use these values to calculate %MMC via the formula <code>1 - (dup_cnt asm / dup_cnt ref)</code>. </p>
<p>Let's try running asmgene on <code>haplotype1</code> and <code>haplotype2</code> from the pre-baked Verkko trio assemblies. </p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="c1">## asmgene</span>
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a><span class="nb">cd</span><span class="w"> </span>~/lra
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>day3_assembly_qc/asmgene
<a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a><span class="nb">cd</span><span class="w"> </span>day3_assembly_qc/asmgene
<a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a><span class="c1"># let&#39;s symlink some of the necessary files</span>
<a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/chm13/chm13v2.0.fa<span class="w"> </span>.
<a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/chm13/CHM13-T2T.cds.fasta<span class="w"> </span>.
<a id="__codelineno-15-8" name="__codelineno-15-8" href="#__codelineno-15-8"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/assembly.haplotype1.fasta<span class="w"> </span>.
<a id="__codelineno-15-9" name="__codelineno-15-9" href="#__codelineno-15-9"></a>ln<span class="w"> </span>-s<span class="w"> </span>/nesi/nobackup/nesi02659/LRA/resources/assemblies/verkko/full/trio/assembly/assembly.haplotype2.fasta<span class="w"> </span>.
</code></pre></div>
</div>
<p>Now that we have our files, we're ready to go. Make a script with the following content and run it in the directory with the appropriate files: </p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a><span class="c1">#SBATCH --job-name      asmgene</span>
<a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a><span class="c1">#SBATCH --cpus-per-task 32</span>
<a id="__codelineno-16-5" name="__codelineno-16-5" href="#__codelineno-16-5"></a><span class="c1">#SBATCH --time          05:00:00</span>
<a id="__codelineno-16-6" name="__codelineno-16-6" href="#__codelineno-16-6"></a><span class="c1">#SBATCH --mem           256G</span>
<a id="__codelineno-16-7" name="__codelineno-16-7" href="#__codelineno-16-7"></a><span class="c1">#SBATCH --output        slurmlogs/%x.%j.log</span>
<a id="__codelineno-16-8" name="__codelineno-16-8" href="#__codelineno-16-8"></a><span class="c1">#SBATCH --error         slurmlogs/%x.%j.err</span>
<a id="__codelineno-16-9" name="__codelineno-16-9" href="#__codelineno-16-9"></a><span class="c1">#SBATCH --partition     milan</span>
<a id="__codelineno-16-10" name="__codelineno-16-10" href="#__codelineno-16-10"></a>
<a id="__codelineno-16-11" name="__codelineno-16-11" href="#__codelineno-16-11"></a><span class="c1">## load modules</span>
<a id="__codelineno-16-12" name="__codelineno-16-12" href="#__codelineno-16-12"></a>module<span class="w"> </span>purge
<a id="__codelineno-16-13" name="__codelineno-16-13" href="#__codelineno-16-13"></a>module<span class="w"> </span>load<span class="w"> </span>minimap2/2.24-GCC-11.3.0
<a id="__codelineno-16-14" name="__codelineno-16-14" href="#__codelineno-16-14"></a>
<a id="__codelineno-16-15" name="__codelineno-16-15" href="#__codelineno-16-15"></a><span class="c1">## run minimap2 on ref, hap1, and hap2</span>
<a id="__codelineno-16-16" name="__codelineno-16-16" href="#__codelineno-16-16"></a>minimap2<span class="w"> </span>-cxsplice:hq<span class="w"> </span>-t32<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-17" name="__codelineno-16-17" href="#__codelineno-16-17"></a><span class="w">    </span>chm13v2.0.fa<span class="w"> </span>CHM13-T2T.cds.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-18" name="__codelineno-16-18" href="#__codelineno-16-18"></a><span class="w">    </span>&gt;<span class="w"> </span>ref.cdna.paf
<a id="__codelineno-16-19" name="__codelineno-16-19" href="#__codelineno-16-19"></a>minimap2<span class="w"> </span>-cxsplice:hq<span class="w"> </span>-t32<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-20" name="__codelineno-16-20" href="#__codelineno-16-20"></a><span class="w">    </span>assembly.haplotype1.fasta<span class="w"> </span>CHM13-T2T.cds.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-21" name="__codelineno-16-21" href="#__codelineno-16-21"></a><span class="w">    </span>&gt;<span class="w"> </span>asm.hap1.cdna.paf
<a id="__codelineno-16-22" name="__codelineno-16-22" href="#__codelineno-16-22"></a>minimap2<span class="w"> </span>-cxsplice:hq<span class="w"> </span>-t32<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-23" name="__codelineno-16-23" href="#__codelineno-16-23"></a><span class="w">    </span>assembly.haplotype2.fasta<span class="w"> </span>CHM13-T2T.cds.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-24" name="__codelineno-16-24" href="#__codelineno-16-24"></a><span class="w">    </span>&gt;<span class="w"> </span>asm.hap2.cdna.paf
<a id="__codelineno-16-25" name="__codelineno-16-25" href="#__codelineno-16-25"></a>
<a id="__codelineno-16-26" name="__codelineno-16-26" href="#__codelineno-16-26"></a><span class="c1">## run asmgene</span>
<a id="__codelineno-16-27" name="__codelineno-16-27" href="#__codelineno-16-27"></a>
<a id="__codelineno-16-28" name="__codelineno-16-28" href="#__codelineno-16-28"></a>k8<span class="w"> </span>/opt/nesi/CS400_centos7_bdw/minimap2/2.24-GCC-11.3.0/bin/paftools.js<span class="w"> </span>asmgene<span class="w"> </span>-a<span class="w"> </span>ref.cdna.paf<span class="w"> </span>asm.hap1.cdna.paf<span class="w"> </span>&gt;<span class="w"> </span>verkko.haplotype1.asmgene.tsv
<a id="__codelineno-16-29" name="__codelineno-16-29" href="#__codelineno-16-29"></a>k8<span class="w"> </span>/opt/nesi/CS400_centos7_bdw/minimap2/2.24-GCC-11.3.0/bin/paftools.js<span class="w"> </span>asmgene<span class="w"> </span>-a<span class="w"> </span>ref.cdna.paf<span class="w"> </span>asm.hap2.cdna.paf<span class="w"> </span>&gt;<span class="w"> </span>verkko.haplotype2.asmgene.tsv
</code></pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Another popular tool for checking genome completeness using gene content is the software Benchmarking Universal Single-Copy Orthologs (BUSCO). This approach uses a set of evolutionarily conserved genes that are expected to be present at single copy for a given taxa, so one could check their genome to see if, for instance, it has all the genes predicted to be necessary for <em>Aves</em> or <em>Vertebrata</em>. This approach is useful if your <em>de novo</em> genome assembly is for a species that does not have a reference genome yet. And it's even faster now with the recently developed tool <em>minibusco</em>!</p>
</div>





                
              </article>
            </div>
          
          
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
            Back to top
          </button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../day2a_assembly/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 4. Assembly" rel="prev">
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                4. Assembly
              </div>
            </div>
          </a>
        
        
          
          <a href="../day3b_assembly_cleanup_annotation/" class="md-footer__link md-footer__link--next" aria-label="Next: 6. Assembly Cleanup &amp; Genome Annotation" rel="next">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                6. Assembly Cleanup & Genome Annotation
              </div>
            </div>
            <div class="md-footer__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Long-read Assembly workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.74e28a9f.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.220ee61c.min.js"></script>
      
        
          <script src="../../javascripts/mathjax.js"></script>
        
      
        
          <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
        
      
        
          <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        
      
    
  <script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>